<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Изучение-и-подготовка-данных-для-BERT" data-toc-modified-id="Изучение-и-подготовка-данных-для-BERT-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Изучение и подготовка данных для BERT</a></span></li><li><span><a href="#Построение-моделей" data-toc-modified-id="Построение-моделей-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Построение моделей</a></span><ul class="toc-item"><li><span><a href="#Модель-логистической-регрессии" data-toc-modified-id="Модель-логистической-регрессии-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Модель логистической регрессии</a></span></li><li><span><a href="#Модель-LightGBM" data-toc-modified-id="Модель-LightGBM-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Модель LightGBM</a></span></li><li><span><a href="#Модель-CatBoost" data-toc-modified-id="Модель-CatBoost-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Модель CatBoost</a></span></li></ul></li><li><span><a href="#Вывод-по-BERT:" data-toc-modified-id="Вывод-по-BERT:-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Вывод по BERT:</a></span></li><li><span><a href="#Подготовка-данных-для-классификации-текстов" data-toc-modified-id="Подготовка-данных-для-классификации-текстов-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Подготовка данных для классификации текстов</a></span></li><li><span><a href="#Построение-моделей" data-toc-modified-id="Построение-моделей-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Построение моделей</a></span><ul class="toc-item"><li><span><a href="#Модель-логистической-регрессии" data-toc-modified-id="Модель-логистической-регрессии-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Модель логистической регрессии</a></span></li><li><span><a href="#Модель-LigthGBM" data-toc-modified-id="Модель-LigthGBM-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Модель LigthGBM</a></span></li><li><span><a href="#Модель-CatBoost" data-toc-modified-id="Модель-CatBoost-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Модель CatBoost</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

## Изучение и подготовка данных для BERT

In [1]:
import numpy as np
import pandas as pd
import torch
import transformers as ppb
from tqdm import notebook
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score

from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier

from sklearn.model_selection import GridSearchCV

import re
import nltk
from nltk.stem import WordNetLemmatizer 
from nltk.corpus import stopwords as nltk_stopwords
from sklearn.feature_extraction.text import TfidfVectorizer

In [2]:
# Код для чтения данных
try:
    df_toxic = pd.read_csv('df_toxic.csv')
except:
    df_toxic= pd.read_csv('https://code.s3.yandex.net/datasets/toxic_comments.csv')
    df_toxic.to_csv('df_toxic.csv', index=False)

In [3]:
# Будем работать только с 1000 строками ради сокращения времени выполнения кода
df_toxic = df_toxic.sample(n=1000, random_state=12345).reset_index(drop=True)

In [4]:
df_toxic

text  toxic
0    Ahh shut the fuck up you douchebag sand nigger...      1
1    "\n\nREPLY: There is no such thing as Texas Co...      0
2    Reply\nHey, you could at least mention Jasenov...      0
3            Thats fine, there is no deadline )   chi?      0
4    "\n\nDYK nomination of Mustarabim\n Hello! You...      0
..                                                 ...    ...
995  Date warriors \n\nHi, Hertz. Nice catch on Sec...      0
996  REDIRECT Talk:2013 Men's World Ice Hockey Cham...      0
997  This article is one of the most well cited art...      0
998  Harry the point you are making is pointless. I...      0
999  "Moreover, ""a translation of the concept x in...      0

[1000 rows x 2 columns]

In [5]:
df_toxic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    1000 non-null   object
 1   toxic   1000 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 15.8+ KB


In [6]:
# Найдем дубликаты
df_toxic[df_toxic.duplicated() == True]

Empty DataFrame
Columns: [text, toxic]
Index: []

In [7]:
# Найдем распределение оценок
df_toxic['toxic'].value_counts()

0    901
1     99
Name: toxic, dtype: int64

In [8]:
# Загрузка предобученной модели DistilBERT
model_class, tokenizer_class, pretrained_weights = (ppb.DistilBertModel,
                                                    ppb.DistilBertTokenizer, 'distilbert-base-uncased')
# Загрузка предобученной модели/токенизатора 
tokenizer = tokenizer_class.from_pretrained(pretrained_weights)
model = model_class.from_pretrained(pretrained_weights)

Some weights of the model checkpoint at distilbert-base-uncased were not used when initializing DistilBertModel: ['vocab_layer_norm.bias', 'vocab_projector.weight', 'vocab_projector.bias', 'vocab_transform.weight', 'vocab_transform.bias', 'vocab_layer_norm.weight']
- This IS expected if you are initializing DistilBertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


In [9]:
# Токенизируем каждый комментарий
tokenized = df_toxic['text'].apply((lambda x:
                                    tokenizer.encode(x, add_special_tokens=True)))

Token indices sequence length is longer than the specified maximum sequence length for this model (862 > 512). Running this sequence through the model will result in indexing errors


In [10]:
# Найдем максимальную длину векторов после токенизации
max_len = 0
for i in tokenized.values:
    if len(i) > max_len:
        max_len = len(i)
# Применим padding к векторам
padded = np.array([i + [0]*(max_len-len(i)) for i in tokenized.values])

In [11]:
# Создадим маску для выделения важных токенов
attention_mask = np.where(padded != 0, 1, 0)

In [12]:
attention_mask.shape

(1000, 4645)

In [13]:
# Ограничим количество токенов
padded_01 = padded[:,:512]
attention_mask_01 = attention_mask[:,:512]

In [14]:
# Сделаем цикл по батчам
batch_size = 100
embeddings = []
for i in notebook.tqdm(range(padded_01.shape[0] // batch_size)):
        # Преобразуем данные в формат тензоров
        batch = torch.LongTensor(padded_01[batch_size*i:batch_size*(i+1)]) 
        attention_mask_batch = torch.LongTensor(attention_mask_01[batch_size*i:batch_size*(i+1)])
        # Укажем, что градиенты не нужны: модель BERT обучать не будем.
        with torch.no_grad():
            batch_embeddings = model(batch, attention_mask=attention_mask_batch)
        # # преобразуем элементы методом numpy() к типу numpy.array
        embeddings.append(batch_embeddings[0][:,0,:].numpy())

  0%|          | 0/10 [00:00<?, ?it/s]

## Построение моделей

In [15]:
# Соберём все эмбеддинги в матрицу признаков
features = np.concatenate(embeddings)
target = df_toxic['toxic']
# Разделим данные на тренировочную и тестовую выборки
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.25, random_state=12345)

### Модель логистической регрессии

In [25]:
model_lr = LogisticRegression(random_state=12345)
result_lr = cross_val_score(model_lr, features_train, target_train,
                                 scoring='f1').mean()
print('F1-мера модели логистической регрессии:', result_lr)

F1-мера модели логистической регрессии: 0.7165217391304347


/opt/anaconda3/envs/practicum/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/opt/anaconda3/envs/practicum/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/

In [26]:
# Проверим F1-меру на тестовой выборке
model_lr.fit(features_train, target_train)
predictions_lr = model_lr.predict(features_test)
print('F1-мера тестовой выборки', f1_score(target_test, predictions_lr))

F1-мера тестовой выборки 0.7599999999999999


/opt/anaconda3/envs/practicum/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### Модель LightGBM

In [18]:
model_lgbm = LGBMClassifier(random_state=12345)
params_lgbm = {'n_estimators' : [100, 200, 500],
               'max_depth' : range(2, 7)}

search_lgbm = GridSearchCV(estimator=model_lgbm, param_grid=params_lgbm, scoring='f1',
                           verbose=3, cv=5)
result_lgbm = search_lgbm.fit(features_train , target_train)

Fitting 5 folds for each of 15 candidates, totalling 75 fits
[CV 1/5] END .....max_depth=2, n_estimators=100;, score=0.571 total time=   0.2s
[CV 2/5] END .....max_depth=2, n_estimators=100;, score=0.783 total time=   0.1s
[CV 3/5] END .....max_depth=2, n_estimators=100;, score=0.500 total time=   0.1s
[CV 4/5] END .....max_depth=2, n_estimators=100;, score=0.667 total time=   0.1s
[CV 5/5] END .....max_depth=2, n_estimators=100;, score=0.692 total time=   0.1s
[CV 1/5] END .....max_depth=2, n_estimators=200;, score=0.636 total time=   0.2s
[CV 2/5] END .....max_depth=2, n_estimators=200;, score=0.727 total time=   0.2s
[CV 3/5] END .....max_depth=2, n_estimators=200;, score=0.500 total time=   0.2s
[CV 4/5] END .....max_depth=2, n_estimators=200;, score=0.600 total time=   0.2s
[CV 5/5] END .....max_depth=2, n_estimators=200;, score=0.692 total time=   0.2s
[CV 1/5] END .....max_depth=2, n_estimators=500;, score=0.636 total time=   0.5s
[CV 2/5] END .....max_depth=2, n_estimators=500;

[CV 2/5] END .....max_depth=6, n_estimators=200;, score=0.667 total time=   0.8s
[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
[CV 3/5] END .....max_depth=6, n_estimators=200;, score=0.476 total time=   0.8s
[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
[CV 4/5] END .....max_depth=6, n_estimators=200;, score=0.571 total time=   0.8s
[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
[CV 5/5] END .....max_depth=6, n_estimators=200;, score=0.720 total time=   0.8s
[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
[CV 1/5] END .....max_depth=6, n_estimators=500;, score=0.600 total time=   0.9s
[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set nu

In [19]:
print('Лучшее значение F1-меры в LightGBM:', result_lgbm.best_score_)
print('Лучшие параметры LightGBM:', result_lgbm.best_params_)

Лучшее значение F1-меры в LightGBM: 0.6426023252110209
Лучшие параметры LightGBM: {'max_depth': 2, 'n_estimators': 100}


In [20]:
# Проверим модель LightGBM на тестовой выборке
best_model_lgbm = search_lgbm.best_estimator_
best_model_lgbm.fit(features_train, target_train)
predictions_lgbm=best_model_lgbm.predict(features_test)
print("F1-мера лучшей модели на тестовой выборке:", f1_score(target_test, predictions_lgbm))

F1-мера лучшей модели на тестовой выборке: 0.6666666666666667


### Модель CatBoost

In [21]:
model_cat = CatBoostClassifier(random_seed=12345)
params_cat = {'iterations' : [100, 300, 500],
              'depth' : range(4, 8)}

search_cat = GridSearchCV(estimator=model_cat, param_grid=params_cat, scoring='f1',
                           verbose=1, cv=3)
result_cat = search_cat.fit(features_train, target_train)

Fitting 3 folds for each of 12 candidates, totalling 36 fits
Learning rate set to 0.063299
0:	learn: 0.6195273	total: 79.9ms	remaining: 7.91s
1:	learn: 0.5661801	total: 90.7ms	remaining: 4.45s
2:	learn: 0.5104177	total: 102ms	remaining: 3.29s
3:	learn: 0.4674025	total: 114ms	remaining: 2.73s
4:	learn: 0.4229842	total: 126ms	remaining: 2.4s
5:	learn: 0.3874266	total: 138ms	remaining: 2.17s
6:	learn: 0.3529163	total: 150ms	remaining: 1.99s
7:	learn: 0.3255369	total: 161ms	remaining: 1.85s
8:	learn: 0.3033351	total: 172ms	remaining: 1.74s
9:	learn: 0.2848788	total: 183ms	remaining: 1.64s
10:	learn: 0.2668043	total: 193ms	remaining: 1.56s
11:	learn: 0.2553160	total: 205ms	remaining: 1.5s
12:	learn: 0.2395211	total: 218ms	remaining: 1.46s
13:	learn: 0.2267390	total: 229ms	remaining: 1.4s
14:	learn: 0.2138451	total: 240ms	remaining: 1.36s
15:	learn: 0.2062809	total: 251ms	remaining: 1.32s
16:	learn: 0.1956071	total: 262ms	remaining: 1.28s
17:	learn: 0.1848688	total: 273ms	remaining: 1.24s
18

65:	learn: 0.0575123	total: 746ms	remaining: 384ms
66:	learn: 0.0561033	total: 758ms	remaining: 373ms
67:	learn: 0.0550902	total: 769ms	remaining: 362ms
68:	learn: 0.0549487	total: 780ms	remaining: 351ms
69:	learn: 0.0533496	total: 791ms	remaining: 339ms
70:	learn: 0.0531841	total: 803ms	remaining: 328ms
71:	learn: 0.0528080	total: 814ms	remaining: 317ms
72:	learn: 0.0514311	total: 825ms	remaining: 305ms
73:	learn: 0.0507822	total: 836ms	remaining: 294ms
74:	learn: 0.0492526	total: 848ms	remaining: 283ms
75:	learn: 0.0482117	total: 860ms	remaining: 271ms
76:	learn: 0.0480418	total: 871ms	remaining: 260ms
77:	learn: 0.0470271	total: 882ms	remaining: 249ms
78:	learn: 0.0468862	total: 895ms	remaining: 238ms
79:	learn: 0.0453431	total: 906ms	remaining: 226ms
80:	learn: 0.0448390	total: 917ms	remaining: 215ms
81:	learn: 0.0432467	total: 928ms	remaining: 204ms
82:	learn: 0.0429340	total: 939ms	remaining: 192ms
83:	learn: 0.0424392	total: 951ms	remaining: 181ms
84:	learn: 0.0414263	total: 964

29:	learn: 0.2720812	total: 341ms	remaining: 3.07s
30:	learn: 0.2648274	total: 353ms	remaining: 3.07s
31:	learn: 0.2591533	total: 365ms	remaining: 3.05s
32:	learn: 0.2538454	total: 376ms	remaining: 3.04s
33:	learn: 0.2475272	total: 387ms	remaining: 3.03s
34:	learn: 0.2414170	total: 398ms	remaining: 3.02s
35:	learn: 0.2351534	total: 409ms	remaining: 3s
36:	learn: 0.2307314	total: 420ms	remaining: 2.99s
37:	learn: 0.2261149	total: 432ms	remaining: 2.98s
38:	learn: 0.2212603	total: 443ms	remaining: 2.96s
39:	learn: 0.2162433	total: 455ms	remaining: 2.96s
40:	learn: 0.2122771	total: 467ms	remaining: 2.95s
41:	learn: 0.2088018	total: 478ms	remaining: 2.94s
42:	learn: 0.2045373	total: 490ms	remaining: 2.93s
43:	learn: 0.2006403	total: 501ms	remaining: 2.91s
44:	learn: 0.1973888	total: 512ms	remaining: 2.9s
45:	learn: 0.1931488	total: 524ms	remaining: 2.89s
46:	learn: 0.1888881	total: 535ms	remaining: 2.88s
47:	learn: 0.1846547	total: 546ms	remaining: 2.87s
48:	learn: 0.1812279	total: 558ms	r

191:	learn: 0.0495564	total: 2.2s	remaining: 1.24s
192:	learn: 0.0491687	total: 2.21s	remaining: 1.23s
193:	learn: 0.0485676	total: 2.22s	remaining: 1.22s
194:	learn: 0.0481899	total: 2.24s	remaining: 1.2s
195:	learn: 0.0478528	total: 2.25s	remaining: 1.19s
196:	learn: 0.0474142	total: 2.26s	remaining: 1.18s
197:	learn: 0.0473494	total: 2.27s	remaining: 1.17s
198:	learn: 0.0470698	total: 2.28s	remaining: 1.16s
199:	learn: 0.0465086	total: 2.29s	remaining: 1.15s
200:	learn: 0.0460585	total: 2.3s	remaining: 1.14s
201:	learn: 0.0459628	total: 2.32s	remaining: 1.12s
202:	learn: 0.0459007	total: 2.33s	remaining: 1.11s
203:	learn: 0.0457771	total: 2.34s	remaining: 1.1s
204:	learn: 0.0455288	total: 2.35s	remaining: 1.09s
205:	learn: 0.0451943	total: 2.36s	remaining: 1.08s
206:	learn: 0.0447410	total: 2.37s	remaining: 1.07s
207:	learn: 0.0446874	total: 2.38s	remaining: 1.05s
208:	learn: 0.0446413	total: 2.4s	remaining: 1.04s
209:	learn: 0.0445051	total: 2.41s	remaining: 1.03s
210:	learn: 0.043

59:	learn: 0.1487383	total: 696ms	remaining: 2.78s
60:	learn: 0.1464228	total: 709ms	remaining: 2.78s
61:	learn: 0.1438930	total: 720ms	remaining: 2.77s
62:	learn: 0.1420744	total: 732ms	remaining: 2.75s
63:	learn: 0.1407646	total: 743ms	remaining: 2.74s
64:	learn: 0.1392600	total: 754ms	remaining: 2.73s
65:	learn: 0.1375005	total: 766ms	remaining: 2.71s
66:	learn: 0.1355190	total: 777ms	remaining: 2.7s
67:	learn: 0.1335862	total: 789ms	remaining: 2.69s
68:	learn: 0.1318109	total: 801ms	remaining: 2.68s
69:	learn: 0.1297884	total: 812ms	remaining: 2.67s
70:	learn: 0.1275858	total: 824ms	remaining: 2.66s
71:	learn: 0.1268696	total: 836ms	remaining: 2.65s
72:	learn: 0.1250676	total: 847ms	remaining: 2.63s
73:	learn: 0.1239534	total: 859ms	remaining: 2.62s
74:	learn: 0.1222091	total: 870ms	remaining: 2.61s
75:	learn: 0.1204798	total: 881ms	remaining: 2.6s
76:	learn: 0.1188045	total: 893ms	remaining: 2.59s
77:	learn: 0.1173305	total: 905ms	remaining: 2.58s
78:	learn: 0.1157003	total: 917ms

218:	learn: 0.0451694	total: 2.55s	remaining: 943ms
219:	learn: 0.0451082	total: 2.56s	remaining: 932ms
220:	learn: 0.0448775	total: 2.57s	remaining: 920ms
221:	learn: 0.0445969	total: 2.58s	remaining: 909ms
222:	learn: 0.0441529	total: 2.6s	remaining: 897ms
223:	learn: 0.0437089	total: 2.61s	remaining: 885ms
224:	learn: 0.0436703	total: 2.62s	remaining: 873ms
225:	learn: 0.0435053	total: 2.63s	remaining: 862ms
226:	learn: 0.0434715	total: 2.64s	remaining: 850ms
227:	learn: 0.0432503	total: 2.65s	remaining: 839ms
228:	learn: 0.0425637	total: 2.67s	remaining: 827ms
229:	learn: 0.0422949	total: 2.68s	remaining: 815ms
230:	learn: 0.0419070	total: 2.69s	remaining: 804ms
231:	learn: 0.0417448	total: 2.7s	remaining: 792ms
232:	learn: 0.0414354	total: 2.71s	remaining: 781ms
233:	learn: 0.0413530	total: 2.73s	remaining: 769ms
234:	learn: 0.0411304	total: 2.74s	remaining: 757ms
235:	learn: 0.0406024	total: 2.75s	remaining: 746ms
236:	learn: 0.0405734	total: 2.76s	remaining: 734ms
237:	learn: 0.

77:	learn: 0.1319302	total: 917ms	remaining: 2.61s
78:	learn: 0.1304006	total: 929ms	remaining: 2.6s
79:	learn: 0.1290814	total: 942ms	remaining: 2.59s
80:	learn: 0.1280904	total: 954ms	remaining: 2.58s
81:	learn: 0.1272732	total: 966ms	remaining: 2.57s
82:	learn: 0.1256240	total: 977ms	remaining: 2.55s
83:	learn: 0.1248110	total: 988ms	remaining: 2.54s
84:	learn: 0.1231540	total: 1000ms	remaining: 2.53s
85:	learn: 0.1211232	total: 1.01s	remaining: 2.52s
86:	learn: 0.1201574	total: 1.02s	remaining: 2.5s
87:	learn: 0.1194795	total: 1.03s	remaining: 2.49s
88:	learn: 0.1183743	total: 1.04s	remaining: 2.48s
89:	learn: 0.1173740	total: 1.06s	remaining: 2.47s
90:	learn: 0.1153812	total: 1.07s	remaining: 2.46s
91:	learn: 0.1138714	total: 1.08s	remaining: 2.45s
92:	learn: 0.1127691	total: 1.09s	remaining: 2.43s
93:	learn: 0.1115011	total: 1.1s	remaining: 2.42s
94:	learn: 0.1104436	total: 1.12s	remaining: 2.41s
95:	learn: 0.1098384	total: 1.13s	remaining: 2.4s
96:	learn: 0.1085879	total: 1.14s	

241:	learn: 0.0408720	total: 2.86s	remaining: 685ms
242:	learn: 0.0407228	total: 2.87s	remaining: 673ms
243:	learn: 0.0404655	total: 2.88s	remaining: 662ms
244:	learn: 0.0399750	total: 2.9s	remaining: 650ms
245:	learn: 0.0397832	total: 2.91s	remaining: 638ms
246:	learn: 0.0395629	total: 2.92s	remaining: 626ms
247:	learn: 0.0395311	total: 2.93s	remaining: 614ms
248:	learn: 0.0390879	total: 2.94s	remaining: 603ms
249:	learn: 0.0390618	total: 2.96s	remaining: 591ms
250:	learn: 0.0388028	total: 2.97s	remaining: 579ms
251:	learn: 0.0386219	total: 2.98s	remaining: 567ms
252:	learn: 0.0382974	total: 2.99s	remaining: 555ms
253:	learn: 0.0381582	total: 3s	remaining: 543ms
254:	learn: 0.0378749	total: 3.01s	remaining: 532ms
255:	learn: 0.0375170	total: 3.02s	remaining: 520ms
256:	learn: 0.0374838	total: 3.04s	remaining: 508ms
257:	learn: 0.0372376	total: 3.05s	remaining: 496ms
258:	learn: 0.0372147	total: 3.06s	remaining: 484ms
259:	learn: 0.0369413	total: 3.07s	remaining: 473ms
260:	learn: 0.03

108:	learn: 0.1382573	total: 1.28s	remaining: 4.61s
109:	learn: 0.1372391	total: 1.29s	remaining: 4.59s
110:	learn: 0.1361683	total: 1.31s	remaining: 4.59s
111:	learn: 0.1349089	total: 1.32s	remaining: 4.57s
112:	learn: 0.1340626	total: 1.33s	remaining: 4.56s
113:	learn: 0.1324790	total: 1.34s	remaining: 4.55s
114:	learn: 0.1309072	total: 1.35s	remaining: 4.54s
115:	learn: 0.1297707	total: 1.37s	remaining: 4.53s
116:	learn: 0.1294647	total: 1.38s	remaining: 4.52s
117:	learn: 0.1281990	total: 1.39s	remaining: 4.51s
118:	learn: 0.1274551	total: 1.4s	remaining: 4.5s
119:	learn: 0.1269017	total: 1.42s	remaining: 4.48s
120:	learn: 0.1257105	total: 1.43s	remaining: 4.47s
121:	learn: 0.1247801	total: 1.44s	remaining: 4.46s
122:	learn: 0.1242199	total: 1.45s	remaining: 4.45s
123:	learn: 0.1230572	total: 1.46s	remaining: 4.43s
124:	learn: 0.1223898	total: 1.47s	remaining: 4.42s
125:	learn: 0.1218341	total: 1.49s	remaining: 4.41s
126:	learn: 0.1203589	total: 1.5s	remaining: 4.4s
127:	learn: 0.11

267:	learn: 0.0572094	total: 3.17s	remaining: 2.75s
268:	learn: 0.0569183	total: 3.19s	remaining: 2.74s
269:	learn: 0.0566756	total: 3.2s	remaining: 2.73s
270:	learn: 0.0563454	total: 3.21s	remaining: 2.71s
271:	learn: 0.0562955	total: 3.22s	remaining: 2.7s
272:	learn: 0.0558823	total: 3.23s	remaining: 2.69s
273:	learn: 0.0556272	total: 3.25s	remaining: 2.68s
274:	learn: 0.0555296	total: 3.26s	remaining: 2.67s
275:	learn: 0.0551053	total: 3.27s	remaining: 2.65s
276:	learn: 0.0547001	total: 3.28s	remaining: 2.64s
277:	learn: 0.0544098	total: 3.29s	remaining: 2.63s
278:	learn: 0.0542418	total: 3.31s	remaining: 2.62s
279:	learn: 0.0541978	total: 3.32s	remaining: 2.61s
280:	learn: 0.0540221	total: 3.33s	remaining: 2.59s
281:	learn: 0.0536454	total: 3.34s	remaining: 2.58s
282:	learn: 0.0536064	total: 3.35s	remaining: 2.57s
283:	learn: 0.0535219	total: 3.37s	remaining: 2.56s
284:	learn: 0.0532057	total: 3.38s	remaining: 2.55s
285:	learn: 0.0531354	total: 3.39s	remaining: 2.54s
286:	learn: 0.

431:	learn: 0.0298628	total: 5.13s	remaining: 808ms
432:	learn: 0.0297251	total: 5.14s	remaining: 796ms
433:	learn: 0.0296713	total: 5.16s	remaining: 784ms
434:	learn: 0.0296417	total: 5.17s	remaining: 772ms
435:	learn: 0.0294215	total: 5.18s	remaining: 760ms
436:	learn: 0.0293055	total: 5.19s	remaining: 749ms
437:	learn: 0.0291625	total: 5.2s	remaining: 737ms
438:	learn: 0.0291362	total: 5.22s	remaining: 725ms
439:	learn: 0.0290864	total: 5.23s	remaining: 713ms
440:	learn: 0.0290067	total: 5.24s	remaining: 701ms
441:	learn: 0.0289308	total: 5.25s	remaining: 689ms
442:	learn: 0.0289177	total: 5.26s	remaining: 677ms
443:	learn: 0.0287388	total: 5.28s	remaining: 666ms
444:	learn: 0.0285521	total: 5.29s	remaining: 654ms
445:	learn: 0.0285328	total: 5.3s	remaining: 642ms
446:	learn: 0.0283873	total: 5.31s	remaining: 630ms
447:	learn: 0.0282104	total: 5.32s	remaining: 618ms
448:	learn: 0.0280551	total: 5.33s	remaining: 606ms
449:	learn: 0.0278839	total: 5.35s	remaining: 594ms
450:	learn: 0.

98:	learn: 0.1447729	total: 1.18s	remaining: 4.8s
99:	learn: 0.1439213	total: 1.2s	remaining: 4.79s
100:	learn: 0.1425067	total: 1.21s	remaining: 4.78s
101:	learn: 0.1413225	total: 1.22s	remaining: 4.77s
102:	learn: 0.1404582	total: 1.23s	remaining: 4.75s
103:	learn: 0.1395389	total: 1.25s	remaining: 4.75s
104:	learn: 0.1379268	total: 1.26s	remaining: 4.73s
105:	learn: 0.1363757	total: 1.27s	remaining: 4.72s
106:	learn: 0.1350943	total: 1.28s	remaining: 4.71s
107:	learn: 0.1336991	total: 1.29s	remaining: 4.7s
108:	learn: 0.1325964	total: 1.31s	remaining: 4.69s
109:	learn: 0.1316559	total: 1.32s	remaining: 4.67s
110:	learn: 0.1301759	total: 1.33s	remaining: 4.66s
111:	learn: 0.1290708	total: 1.34s	remaining: 4.65s
112:	learn: 0.1282487	total: 1.35s	remaining: 4.63s
113:	learn: 0.1270243	total: 1.36s	remaining: 4.62s
114:	learn: 0.1256634	total: 1.38s	remaining: 4.61s
115:	learn: 0.1246800	total: 1.39s	remaining: 4.6s
116:	learn: 0.1239773	total: 1.4s	remaining: 4.59s
117:	learn: 0.12290

268:	learn: 0.0619672	total: 3.23s	remaining: 2.77s
269:	learn: 0.0615954	total: 3.24s	remaining: 2.76s
270:	learn: 0.0612609	total: 3.25s	remaining: 2.75s
271:	learn: 0.0612174	total: 3.27s	remaining: 2.74s
272:	learn: 0.0608874	total: 3.28s	remaining: 2.73s
273:	learn: 0.0608241	total: 3.29s	remaining: 2.71s
274:	learn: 0.0605718	total: 3.3s	remaining: 2.7s
275:	learn: 0.0600829	total: 3.31s	remaining: 2.69s
276:	learn: 0.0597190	total: 3.33s	remaining: 2.68s
277:	learn: 0.0592841	total: 3.34s	remaining: 2.67s
278:	learn: 0.0592350	total: 3.35s	remaining: 2.65s
279:	learn: 0.0589502	total: 3.36s	remaining: 2.64s
280:	learn: 0.0584668	total: 3.37s	remaining: 2.63s
281:	learn: 0.0580849	total: 3.38s	remaining: 2.62s
282:	learn: 0.0578548	total: 3.4s	remaining: 2.6s
283:	learn: 0.0577848	total: 3.41s	remaining: 2.59s
284:	learn: 0.0574345	total: 3.42s	remaining: 2.58s
285:	learn: 0.0573959	total: 3.43s	remaining: 2.57s
286:	learn: 0.0570201	total: 3.44s	remaining: 2.56s
287:	learn: 0.05

436:	learn: 0.0336514	total: 5.3s	remaining: 765ms
437:	learn: 0.0336147	total: 5.32s	remaining: 753ms
438:	learn: 0.0335828	total: 5.33s	remaining: 741ms
439:	learn: 0.0333786	total: 5.35s	remaining: 729ms
440:	learn: 0.0333628	total: 5.36s	remaining: 717ms
441:	learn: 0.0333463	total: 5.38s	remaining: 706ms
442:	learn: 0.0333286	total: 5.39s	remaining: 694ms
443:	learn: 0.0331065	total: 5.4s	remaining: 682ms
444:	learn: 0.0329586	total: 5.42s	remaining: 670ms
445:	learn: 0.0327348	total: 5.43s	remaining: 658ms
446:	learn: 0.0326712	total: 5.45s	remaining: 646ms
447:	learn: 0.0325476	total: 5.46s	remaining: 634ms
448:	learn: 0.0324776	total: 5.47s	remaining: 622ms
449:	learn: 0.0324401	total: 5.49s	remaining: 610ms
450:	learn: 0.0324209	total: 5.5s	remaining: 598ms
451:	learn: 0.0324052	total: 5.51s	remaining: 586ms
452:	learn: 0.0321821	total: 5.53s	remaining: 574ms
453:	learn: 0.0321645	total: 5.54s	remaining: 562ms
454:	learn: 0.0319163	total: 5.56s	remaining: 550ms
455:	learn: 0.0

97:	learn: 0.1624102	total: 1.23s	remaining: 5.04s
98:	learn: 0.1608704	total: 1.24s	remaining: 5.03s
99:	learn: 0.1596910	total: 1.25s	remaining: 5.02s
100:	learn: 0.1581726	total: 1.27s	remaining: 5s
101:	learn: 0.1571169	total: 1.28s	remaining: 4.99s
102:	learn: 0.1562230	total: 1.29s	remaining: 4.98s
103:	learn: 0.1549041	total: 1.31s	remaining: 4.98s
104:	learn: 0.1537792	total: 1.32s	remaining: 4.97s
105:	learn: 0.1524143	total: 1.33s	remaining: 4.96s
106:	learn: 0.1512280	total: 1.35s	remaining: 4.95s
107:	learn: 0.1505423	total: 1.36s	remaining: 4.94s
108:	learn: 0.1490938	total: 1.38s	remaining: 4.95s
109:	learn: 0.1481929	total: 1.39s	remaining: 4.94s
110:	learn: 0.1470712	total: 1.41s	remaining: 4.93s
111:	learn: 0.1462710	total: 1.42s	remaining: 4.92s
112:	learn: 0.1453742	total: 1.43s	remaining: 4.91s
113:	learn: 0.1442939	total: 1.45s	remaining: 4.89s
114:	learn: 0.1430163	total: 1.46s	remaining: 4.89s
115:	learn: 0.1421565	total: 1.47s	remaining: 4.88s
116:	learn: 0.1408

264:	learn: 0.0673852	total: 3.44s	remaining: 3.05s
265:	learn: 0.0672227	total: 3.45s	remaining: 3.04s
266:	learn: 0.0669116	total: 3.47s	remaining: 3.02s
267:	learn: 0.0665267	total: 3.48s	remaining: 3.01s
268:	learn: 0.0661091	total: 3.5s	remaining: 3s
269:	learn: 0.0660652	total: 3.51s	remaining: 2.99s
270:	learn: 0.0657240	total: 3.52s	remaining: 2.98s
271:	learn: 0.0655264	total: 3.54s	remaining: 2.96s
272:	learn: 0.0651045	total: 3.55s	remaining: 2.95s
273:	learn: 0.0650792	total: 3.56s	remaining: 2.94s
274:	learn: 0.0647610	total: 3.58s	remaining: 2.93s
275:	learn: 0.0644174	total: 3.59s	remaining: 2.92s
276:	learn: 0.0642602	total: 3.61s	remaining: 2.9s
277:	learn: 0.0639804	total: 3.62s	remaining: 2.89s
278:	learn: 0.0638650	total: 3.63s	remaining: 2.88s
279:	learn: 0.0636060	total: 3.65s	remaining: 2.87s
280:	learn: 0.0633394	total: 3.66s	remaining: 2.85s
281:	learn: 0.0630174	total: 3.67s	remaining: 2.84s
282:	learn: 0.0627357	total: 3.69s	remaining: 2.83s
283:	learn: 0.062

424:	learn: 0.0348686	total: 5.5s	remaining: 971ms
425:	learn: 0.0347665	total: 5.51s	remaining: 958ms
426:	learn: 0.0346597	total: 5.53s	remaining: 945ms
427:	learn: 0.0345501	total: 5.54s	remaining: 931ms
428:	learn: 0.0343290	total: 5.55s	remaining: 919ms
429:	learn: 0.0341753	total: 5.56s	remaining: 906ms
430:	learn: 0.0340669	total: 5.57s	remaining: 892ms
431:	learn: 0.0339667	total: 5.59s	remaining: 879ms
432:	learn: 0.0338845	total: 5.6s	remaining: 866ms
433:	learn: 0.0337867	total: 5.61s	remaining: 853ms
434:	learn: 0.0336541	total: 5.62s	remaining: 840ms
435:	learn: 0.0336377	total: 5.64s	remaining: 827ms
436:	learn: 0.0335255	total: 5.65s	remaining: 814ms
437:	learn: 0.0333970	total: 5.66s	remaining: 801ms
438:	learn: 0.0332433	total: 5.67s	remaining: 789ms
439:	learn: 0.0331890	total: 5.69s	remaining: 776ms
440:	learn: 0.0331019	total: 5.7s	remaining: 763ms
441:	learn: 0.0330105	total: 5.72s	remaining: 750ms
442:	learn: 0.0329166	total: 5.73s	remaining: 738ms
443:	learn: 0.0

85:	learn: 0.0267464	total: 1.69s	remaining: 275ms
86:	learn: 0.0260902	total: 1.71s	remaining: 255ms
87:	learn: 0.0255925	total: 1.73s	remaining: 235ms
88:	learn: 0.0249872	total: 1.75s	remaining: 216ms
89:	learn: 0.0245914	total: 1.76s	remaining: 196ms
90:	learn: 0.0244272	total: 1.78s	remaining: 176ms
91:	learn: 0.0243491	total: 1.8s	remaining: 157ms
92:	learn: 0.0238003	total: 1.82s	remaining: 137ms
93:	learn: 0.0236392	total: 1.84s	remaining: 117ms
94:	learn: 0.0231149	total: 1.86s	remaining: 97.8ms
95:	learn: 0.0228764	total: 1.88s	remaining: 78.2ms
96:	learn: 0.0223194	total: 1.9s	remaining: 58.7ms
97:	learn: 0.0219195	total: 1.92s	remaining: 39.1ms
98:	learn: 0.0217730	total: 1.94s	remaining: 19.5ms
99:	learn: 0.0213031	total: 1.96s	remaining: 0us
Learning rate set to 0.063299
0:	learn: 0.6276405	total: 31.4ms	remaining: 3.1s
1:	learn: 0.5602422	total: 48.9ms	remaining: 2.4s
2:	learn: 0.5149615	total: 66.6ms	remaining: 2.15s
3:	learn: 0.4680644	total: 85.1ms	remaining: 2.04s
4:

47:	learn: 0.0662087	total: 968ms	remaining: 1.05s
48:	learn: 0.0643612	total: 988ms	remaining: 1.03s
49:	learn: 0.0640090	total: 1.01s	remaining: 1.01s
50:	learn: 0.0625300	total: 1.02s	remaining: 985ms
51:	learn: 0.0609259	total: 1.04s	remaining: 964ms
52:	learn: 0.0605964	total: 1.06s	remaining: 944ms
53:	learn: 0.0591745	total: 1.08s	remaining: 923ms
54:	learn: 0.0571678	total: 1.1s	remaining: 902ms
55:	learn: 0.0558195	total: 1.12s	remaining: 881ms
56:	learn: 0.0546742	total: 1.14s	remaining: 860ms
57:	learn: 0.0534350	total: 1.16s	remaining: 839ms
58:	learn: 0.0531321	total: 1.18s	remaining: 818ms
59:	learn: 0.0516826	total: 1.2s	remaining: 798ms
60:	learn: 0.0506007	total: 1.22s	remaining: 777ms
61:	learn: 0.0500590	total: 1.23s	remaining: 756ms
62:	learn: 0.0486297	total: 1.25s	remaining: 737ms
63:	learn: 0.0476400	total: 1.28s	remaining: 719ms
64:	learn: 0.0464019	total: 1.3s	remaining: 700ms
65:	learn: 0.0450414	total: 1.32s	remaining: 681ms
66:	learn: 0.0443281	total: 1.34s	

112:	learn: 0.0736504	total: 2.3s	remaining: 3.81s
113:	learn: 0.0729723	total: 2.32s	remaining: 3.79s
114:	learn: 0.0716450	total: 2.34s	remaining: 3.77s
115:	learn: 0.0713294	total: 2.36s	remaining: 3.74s
116:	learn: 0.0708626	total: 2.38s	remaining: 3.72s
117:	learn: 0.0706790	total: 2.4s	remaining: 3.7s
118:	learn: 0.0698127	total: 2.42s	remaining: 3.68s
119:	learn: 0.0686707	total: 2.44s	remaining: 3.65s
120:	learn: 0.0676241	total: 2.45s	remaining: 3.63s
121:	learn: 0.0665767	total: 2.47s	remaining: 3.61s
122:	learn: 0.0662467	total: 2.49s	remaining: 3.59s
123:	learn: 0.0657261	total: 2.51s	remaining: 3.56s
124:	learn: 0.0651518	total: 2.53s	remaining: 3.54s
125:	learn: 0.0648870	total: 2.55s	remaining: 3.52s
126:	learn: 0.0642047	total: 2.57s	remaining: 3.5s
127:	learn: 0.0636800	total: 2.59s	remaining: 3.48s
128:	learn: 0.0629535	total: 2.61s	remaining: 3.46s
129:	learn: 0.0624878	total: 2.63s	remaining: 3.43s
130:	learn: 0.0616285	total: 2.65s	remaining: 3.41s
131:	learn: 0.06

274:	learn: 0.0217443	total: 5.48s	remaining: 498ms
275:	learn: 0.0215271	total: 5.5s	remaining: 478ms
276:	learn: 0.0212819	total: 5.51s	remaining: 458ms
277:	learn: 0.0212609	total: 5.53s	remaining: 438ms
278:	learn: 0.0210489	total: 5.55s	remaining: 418ms
279:	learn: 0.0209953	total: 5.57s	remaining: 398ms
280:	learn: 0.0209499	total: 5.59s	remaining: 378ms
281:	learn: 0.0208118	total: 5.61s	remaining: 358ms
282:	learn: 0.0207800	total: 5.63s	remaining: 338ms
283:	learn: 0.0207344	total: 5.65s	remaining: 318ms
284:	learn: 0.0205405	total: 5.67s	remaining: 298ms
285:	learn: 0.0203327	total: 5.69s	remaining: 278ms
286:	learn: 0.0201707	total: 5.71s	remaining: 259ms
287:	learn: 0.0201485	total: 5.73s	remaining: 239ms
288:	learn: 0.0200193	total: 5.75s	remaining: 219ms
289:	learn: 0.0198176	total: 5.76s	remaining: 199ms
290:	learn: 0.0196285	total: 5.78s	remaining: 179ms
291:	learn: 0.0194749	total: 5.8s	remaining: 159ms
292:	learn: 0.0193206	total: 5.82s	remaining: 139ms
293:	learn: 0.

142:	learn: 0.0566346	total: 2.86s	remaining: 3.14s
143:	learn: 0.0562947	total: 2.88s	remaining: 3.12s
144:	learn: 0.0560237	total: 2.9s	remaining: 3.1s
145:	learn: 0.0551151	total: 2.92s	remaining: 3.08s
146:	learn: 0.0546591	total: 2.93s	remaining: 3.05s
147:	learn: 0.0542682	total: 2.95s	remaining: 3.03s
148:	learn: 0.0540023	total: 2.97s	remaining: 3.01s
149:	learn: 0.0535354	total: 2.99s	remaining: 2.99s
150:	learn: 0.0531107	total: 3.01s	remaining: 2.97s
151:	learn: 0.0530241	total: 3.03s	remaining: 2.95s
152:	learn: 0.0527519	total: 3.05s	remaining: 2.93s
153:	learn: 0.0526780	total: 3.07s	remaining: 2.91s
154:	learn: 0.0520110	total: 3.09s	remaining: 2.89s
155:	learn: 0.0517469	total: 3.11s	remaining: 2.87s
156:	learn: 0.0513415	total: 3.13s	remaining: 2.85s
157:	learn: 0.0510665	total: 3.15s	remaining: 2.83s
158:	learn: 0.0501226	total: 3.17s	remaining: 2.81s
159:	learn: 0.0499822	total: 3.19s	remaining: 2.79s
160:	learn: 0.0494584	total: 3.2s	remaining: 2.77s
161:	learn: 0.0

4:	learn: 0.5787253	total: 105ms	remaining: 6.21s
5:	learn: 0.5576699	total: 125ms	remaining: 6.11s
6:	learn: 0.5388104	total: 144ms	remaining: 6.03s
7:	learn: 0.5182846	total: 163ms	remaining: 5.93s
8:	learn: 0.5012423	total: 181ms	remaining: 5.86s
9:	learn: 0.4839825	total: 200ms	remaining: 5.81s
10:	learn: 0.4693863	total: 219ms	remaining: 5.75s
11:	learn: 0.4557390	total: 238ms	remaining: 5.71s
12:	learn: 0.4435075	total: 257ms	remaining: 5.68s
13:	learn: 0.4295054	total: 276ms	remaining: 5.64s
14:	learn: 0.4181592	total: 295ms	remaining: 5.61s
15:	learn: 0.4042907	total: 314ms	remaining: 5.58s
16:	learn: 0.3923584	total: 335ms	remaining: 5.58s
17:	learn: 0.3823323	total: 354ms	remaining: 5.54s
18:	learn: 0.3742926	total: 373ms	remaining: 5.52s
19:	learn: 0.3638663	total: 392ms	remaining: 5.48s
20:	learn: 0.3530935	total: 410ms	remaining: 5.45s
21:	learn: 0.3439184	total: 429ms	remaining: 5.42s
22:	learn: 0.3322296	total: 448ms	remaining: 5.4s
23:	learn: 0.3216601	total: 467ms	rema

167:	learn: 0.0507766	total: 3.29s	remaining: 2.58s
168:	learn: 0.0502560	total: 3.31s	remaining: 2.56s
169:	learn: 0.0500299	total: 3.33s	remaining: 2.54s
170:	learn: 0.0494051	total: 3.35s	remaining: 2.53s
171:	learn: 0.0488979	total: 3.37s	remaining: 2.51s
172:	learn: 0.0484313	total: 3.39s	remaining: 2.49s
173:	learn: 0.0480423	total: 3.41s	remaining: 2.47s
174:	learn: 0.0478779	total: 3.43s	remaining: 2.45s
175:	learn: 0.0475182	total: 3.44s	remaining: 2.43s
176:	learn: 0.0472581	total: 3.46s	remaining: 2.41s
177:	learn: 0.0468546	total: 3.48s	remaining: 2.39s
178:	learn: 0.0465544	total: 3.5s	remaining: 2.37s
179:	learn: 0.0460269	total: 3.52s	remaining: 2.35s
180:	learn: 0.0456083	total: 3.54s	remaining: 2.33s
181:	learn: 0.0453903	total: 3.56s	remaining: 2.31s
182:	learn: 0.0450873	total: 3.58s	remaining: 2.29s
183:	learn: 0.0447907	total: 3.6s	remaining: 2.27s
184:	learn: 0.0444613	total: 3.62s	remaining: 2.25s
185:	learn: 0.0441574	total: 3.64s	remaining: 2.23s
186:	learn: 0.

29:	learn: 0.3564390	total: 589ms	remaining: 9.22s
30:	learn: 0.3511860	total: 610ms	remaining: 9.23s
31:	learn: 0.3446779	total: 629ms	remaining: 9.2s
32:	learn: 0.3381974	total: 648ms	remaining: 9.17s
33:	learn: 0.3327730	total: 667ms	remaining: 9.14s
34:	learn: 0.3261096	total: 686ms	remaining: 9.12s
35:	learn: 0.3199983	total: 705ms	remaining: 9.09s
36:	learn: 0.3147410	total: 724ms	remaining: 9.06s
37:	learn: 0.3099709	total: 743ms	remaining: 9.03s
38:	learn: 0.3050783	total: 762ms	remaining: 9.01s
39:	learn: 0.3010081	total: 781ms	remaining: 8.98s
40:	learn: 0.2949298	total: 800ms	remaining: 8.96s
41:	learn: 0.2900595	total: 819ms	remaining: 8.93s
42:	learn: 0.2846303	total: 839ms	remaining: 8.92s
43:	learn: 0.2798808	total: 859ms	remaining: 8.9s
44:	learn: 0.2753800	total: 878ms	remaining: 8.88s
45:	learn: 0.2708018	total: 897ms	remaining: 8.85s
46:	learn: 0.2668282	total: 916ms	remaining: 8.83s
47:	learn: 0.2619660	total: 935ms	remaining: 8.8s
48:	learn: 0.2574270	total: 954ms	

195:	learn: 0.0645655	total: 4s	remaining: 6.21s
196:	learn: 0.0641784	total: 4.02s	remaining: 6.19s
197:	learn: 0.0636921	total: 4.04s	remaining: 6.17s
198:	learn: 0.0633059	total: 4.06s	remaining: 6.14s
199:	learn: 0.0630012	total: 4.08s	remaining: 6.12s
200:	learn: 0.0627107	total: 4.1s	remaining: 6.1s
201:	learn: 0.0623827	total: 4.12s	remaining: 6.08s
202:	learn: 0.0620310	total: 4.14s	remaining: 6.05s
203:	learn: 0.0619373	total: 4.16s	remaining: 6.03s
204:	learn: 0.0614111	total: 4.18s	remaining: 6.01s
205:	learn: 0.0609123	total: 4.2s	remaining: 5.99s
206:	learn: 0.0603983	total: 4.21s	remaining: 5.97s
207:	learn: 0.0599775	total: 4.23s	remaining: 5.94s
208:	learn: 0.0596234	total: 4.25s	remaining: 5.92s
209:	learn: 0.0591750	total: 4.27s	remaining: 5.9s
210:	learn: 0.0587770	total: 4.29s	remaining: 5.88s
211:	learn: 0.0586802	total: 4.32s	remaining: 5.86s
212:	learn: 0.0585701	total: 4.34s	remaining: 5.84s
213:	learn: 0.0583766	total: 4.36s	remaining: 5.82s
214:	learn: 0.05814

354:	learn: 0.0289646	total: 7.18s	remaining: 2.93s
355:	learn: 0.0288262	total: 7.2s	remaining: 2.91s
356:	learn: 0.0287150	total: 7.22s	remaining: 2.89s
357:	learn: 0.0286277	total: 7.24s	remaining: 2.87s
358:	learn: 0.0285010	total: 7.26s	remaining: 2.85s
359:	learn: 0.0283087	total: 7.28s	remaining: 2.83s
360:	learn: 0.0282332	total: 7.3s	remaining: 2.81s
361:	learn: 0.0280476	total: 7.32s	remaining: 2.79s
362:	learn: 0.0278747	total: 7.34s	remaining: 2.77s
363:	learn: 0.0277781	total: 7.36s	remaining: 2.75s
364:	learn: 0.0275907	total: 7.37s	remaining: 2.73s
365:	learn: 0.0275124	total: 7.39s	remaining: 2.71s
366:	learn: 0.0274872	total: 7.42s	remaining: 2.69s
367:	learn: 0.0273274	total: 7.43s	remaining: 2.67s
368:	learn: 0.0271753	total: 7.45s	remaining: 2.65s
369:	learn: 0.0271413	total: 7.47s	remaining: 2.63s
370:	learn: 0.0270543	total: 7.49s	remaining: 2.6s
371:	learn: 0.0269918	total: 7.51s	remaining: 2.58s
372:	learn: 0.0268235	total: 7.53s	remaining: 2.56s
373:	learn: 0.0

22:	learn: 0.4148187	total: 478ms	remaining: 9.91s
23:	learn: 0.4042266	total: 498ms	remaining: 9.87s
24:	learn: 0.3968198	total: 517ms	remaining: 9.82s
25:	learn: 0.3882783	total: 536ms	remaining: 9.77s
26:	learn: 0.3820821	total: 555ms	remaining: 9.73s
27:	learn: 0.3749509	total: 575ms	remaining: 9.68s
28:	learn: 0.3683671	total: 594ms	remaining: 9.64s
29:	learn: 0.3611252	total: 613ms	remaining: 9.61s
30:	learn: 0.3515329	total: 632ms	remaining: 9.56s
31:	learn: 0.3444256	total: 652ms	remaining: 9.53s
32:	learn: 0.3358690	total: 671ms	remaining: 9.49s
33:	learn: 0.3288024	total: 691ms	remaining: 9.47s
34:	learn: 0.3238560	total: 710ms	remaining: 9.44s
35:	learn: 0.3171509	total: 730ms	remaining: 9.41s
36:	learn: 0.3115389	total: 749ms	remaining: 9.38s
37:	learn: 0.3044097	total: 769ms	remaining: 9.34s
38:	learn: 0.2992929	total: 787ms	remaining: 9.3s
39:	learn: 0.2933961	total: 806ms	remaining: 9.27s
40:	learn: 0.2886001	total: 825ms	remaining: 9.24s
41:	learn: 0.2828491	total: 844m

182:	learn: 0.0713193	total: 3.69s	remaining: 6.39s
183:	learn: 0.0709064	total: 3.71s	remaining: 6.38s
184:	learn: 0.0705206	total: 3.73s	remaining: 6.36s
185:	learn: 0.0699984	total: 3.75s	remaining: 6.34s
186:	learn: 0.0696595	total: 3.77s	remaining: 6.31s
187:	learn: 0.0693433	total: 3.79s	remaining: 6.29s
188:	learn: 0.0691984	total: 3.81s	remaining: 6.27s
189:	learn: 0.0688432	total: 3.83s	remaining: 6.25s
190:	learn: 0.0684447	total: 3.85s	remaining: 6.23s
191:	learn: 0.0677925	total: 3.87s	remaining: 6.21s
192:	learn: 0.0672485	total: 3.89s	remaining: 6.19s
193:	learn: 0.0666074	total: 3.92s	remaining: 6.17s
194:	learn: 0.0662801	total: 3.94s	remaining: 6.16s
195:	learn: 0.0662068	total: 3.96s	remaining: 6.13s
196:	learn: 0.0661136	total: 3.98s	remaining: 6.11s
197:	learn: 0.0654471	total: 4s	remaining: 6.1s
198:	learn: 0.0651847	total: 4.02s	remaining: 6.08s
199:	learn: 0.0648908	total: 4.04s	remaining: 6.05s
200:	learn: 0.0648379	total: 4.06s	remaining: 6.04s
201:	learn: 0.06

341:	learn: 0.0340770	total: 6.86s	remaining: 3.17s
342:	learn: 0.0339490	total: 6.88s	remaining: 3.15s
343:	learn: 0.0337836	total: 6.9s	remaining: 3.13s
344:	learn: 0.0337062	total: 6.92s	remaining: 3.11s
345:	learn: 0.0336212	total: 6.94s	remaining: 3.09s
346:	learn: 0.0334637	total: 6.96s	remaining: 3.07s
347:	learn: 0.0332706	total: 6.98s	remaining: 3.05s
348:	learn: 0.0331361	total: 7s	remaining: 3.03s
349:	learn: 0.0329610	total: 7.02s	remaining: 3.01s
350:	learn: 0.0328529	total: 7.04s	remaining: 2.99s
351:	learn: 0.0328343	total: 7.05s	remaining: 2.97s
352:	learn: 0.0326602	total: 7.08s	remaining: 2.95s
353:	learn: 0.0324733	total: 7.1s	remaining: 2.93s
354:	learn: 0.0323375	total: 7.12s	remaining: 2.91s
355:	learn: 0.0320725	total: 7.13s	remaining: 2.88s
356:	learn: 0.0319699	total: 7.15s	remaining: 2.87s
357:	learn: 0.0319286	total: 7.17s	remaining: 2.85s
358:	learn: 0.0317583	total: 7.19s	remaining: 2.82s
359:	learn: 0.0317438	total: 7.21s	remaining: 2.8s
360:	learn: 0.0315

Learning rate set to 0.014469
0:	learn: 0.6780679	total: 31.4ms	remaining: 15.7s
1:	learn: 0.6604597	total: 49.6ms	remaining: 12.3s
2:	learn: 0.6460175	total: 67.9ms	remaining: 11.2s
3:	learn: 0.6328664	total: 86.1ms	remaining: 10.7s
4:	learn: 0.6179824	total: 105ms	remaining: 10.4s
5:	learn: 0.6028276	total: 124ms	remaining: 10.2s
6:	learn: 0.5893329	total: 143ms	remaining: 10.1s
7:	learn: 0.5744333	total: 162ms	remaining: 9.98s
8:	learn: 0.5616885	total: 182ms	remaining: 9.91s
9:	learn: 0.5489264	total: 202ms	remaining: 9.92s
10:	learn: 0.5375175	total: 224ms	remaining: 9.97s
11:	learn: 0.5268780	total: 244ms	remaining: 9.91s
12:	learn: 0.5167801	total: 264ms	remaining: 9.87s
13:	learn: 0.5053046	total: 283ms	remaining: 9.81s
14:	learn: 0.4955682	total: 302ms	remaining: 9.77s
15:	learn: 0.4838504	total: 321ms	remaining: 9.7s
16:	learn: 0.4747283	total: 340ms	remaining: 9.66s
17:	learn: 0.4657203	total: 359ms	remaining: 9.62s
18:	learn: 0.4581101	total: 379ms	remaining: 9.59s
19:	lear

163:	learn: 0.0897520	total: 3.2s	remaining: 6.55s
164:	learn: 0.0891132	total: 3.22s	remaining: 6.53s
165:	learn: 0.0884015	total: 3.24s	remaining: 6.51s
166:	learn: 0.0876831	total: 3.26s	remaining: 6.49s
167:	learn: 0.0875574	total: 3.28s	remaining: 6.47s
168:	learn: 0.0869149	total: 3.29s	remaining: 6.45s
169:	learn: 0.0862635	total: 3.31s	remaining: 6.43s
170:	learn: 0.0855658	total: 3.33s	remaining: 6.42s
171:	learn: 0.0849381	total: 3.35s	remaining: 6.39s
172:	learn: 0.0842503	total: 3.37s	remaining: 6.38s
173:	learn: 0.0838823	total: 3.39s	remaining: 6.35s
174:	learn: 0.0836318	total: 3.41s	remaining: 6.34s
175:	learn: 0.0832145	total: 3.43s	remaining: 6.32s
176:	learn: 0.0827044	total: 3.45s	remaining: 6.3s
177:	learn: 0.0822716	total: 3.47s	remaining: 6.28s
178:	learn: 0.0819254	total: 3.49s	remaining: 6.26s
179:	learn: 0.0811067	total: 3.51s	remaining: 6.24s
180:	learn: 0.0807278	total: 3.53s	remaining: 6.22s
181:	learn: 0.0804965	total: 3.55s	remaining: 6.2s
182:	learn: 0.0

328:	learn: 0.0378416	total: 6.42s	remaining: 3.34s
329:	learn: 0.0376107	total: 6.45s	remaining: 3.32s
330:	learn: 0.0374574	total: 6.47s	remaining: 3.3s
331:	learn: 0.0373210	total: 6.49s	remaining: 3.28s
332:	learn: 0.0373006	total: 6.51s	remaining: 3.26s
333:	learn: 0.0371798	total: 6.53s	remaining: 3.24s
334:	learn: 0.0371101	total: 6.55s	remaining: 3.22s
335:	learn: 0.0370286	total: 6.57s	remaining: 3.21s
336:	learn: 0.0368247	total: 6.58s	remaining: 3.19s
337:	learn: 0.0367491	total: 6.61s	remaining: 3.17s
338:	learn: 0.0365356	total: 6.62s	remaining: 3.15s
339:	learn: 0.0364436	total: 6.64s	remaining: 3.13s
340:	learn: 0.0363719	total: 6.67s	remaining: 3.11s
341:	learn: 0.0361340	total: 6.68s	remaining: 3.09s
342:	learn: 0.0359856	total: 6.7s	remaining: 3.07s
343:	learn: 0.0358143	total: 6.72s	remaining: 3.05s
344:	learn: 0.0356374	total: 6.74s	remaining: 3.03s
345:	learn: 0.0354767	total: 6.76s	remaining: 3.01s
346:	learn: 0.0354011	total: 6.78s	remaining: 2.99s
347:	learn: 0.

493:	learn: 0.0198638	total: 9.66s	remaining: 117ms
494:	learn: 0.0197256	total: 9.68s	remaining: 97.8ms
495:	learn: 0.0196311	total: 9.7s	remaining: 78.3ms
496:	learn: 0.0195183	total: 9.72s	remaining: 58.7ms
497:	learn: 0.0194059	total: 9.74s	remaining: 39.1ms
498:	learn: 0.0193262	total: 9.76s	remaining: 19.6ms
499:	learn: 0.0192938	total: 9.78s	remaining: 0us
Learning rate set to 0.063299
0:	learn: 0.6216180	total: 54ms	remaining: 5.34s
1:	learn: 0.5569407	total: 84.2ms	remaining: 4.13s
2:	learn: 0.5013928	total: 114ms	remaining: 3.68s
3:	learn: 0.4688703	total: 144ms	remaining: 3.47s
4:	learn: 0.4247624	total: 175ms	remaining: 3.32s
5:	learn: 0.3847068	total: 206ms	remaining: 3.22s
6:	learn: 0.3491074	total: 236ms	remaining: 3.13s
7:	learn: 0.3193759	total: 267ms	remaining: 3.07s
8:	learn: 0.2956843	total: 298ms	remaining: 3.01s
9:	learn: 0.2727784	total: 329ms	remaining: 2.96s
10:	learn: 0.2545053	total: 364ms	remaining: 2.94s
11:	learn: 0.2369696	total: 395ms	remaining: 2.9s
12:

57:	learn: 0.0417390	total: 1.91s	remaining: 1.38s
58:	learn: 0.0406666	total: 1.94s	remaining: 1.35s
59:	learn: 0.0390106	total: 1.97s	remaining: 1.31s
60:	learn: 0.0384032	total: 2s	remaining: 1.28s
61:	learn: 0.0376082	total: 2.03s	remaining: 1.25s
62:	learn: 0.0374584	total: 2.07s	remaining: 1.21s
63:	learn: 0.0364630	total: 2.1s	remaining: 1.18s
64:	learn: 0.0357402	total: 2.13s	remaining: 1.15s
65:	learn: 0.0346896	total: 2.16s	remaining: 1.11s
66:	learn: 0.0337234	total: 2.19s	remaining: 1.08s
67:	learn: 0.0329887	total: 2.23s	remaining: 1.05s
68:	learn: 0.0319320	total: 2.26s	remaining: 1.01s
69:	learn: 0.0311002	total: 2.29s	remaining: 980ms
70:	learn: 0.0304428	total: 2.32s	remaining: 948ms
71:	learn: 0.0302222	total: 2.35s	remaining: 915ms
72:	learn: 0.0299339	total: 2.39s	remaining: 883ms
73:	learn: 0.0292096	total: 2.42s	remaining: 852ms
74:	learn: 0.0285660	total: 2.46s	remaining: 821ms
75:	learn: 0.0283461	total: 2.5s	remaining: 790ms
76:	learn: 0.0275153	total: 2.54s	re

21:	learn: 0.3208172	total: 775ms	remaining: 9.79s
22:	learn: 0.3127745	total: 813ms	remaining: 9.79s
23:	learn: 0.3030155	total: 851ms	remaining: 9.79s
24:	learn: 0.2944731	total: 885ms	remaining: 9.73s
25:	learn: 0.2845189	total: 917ms	remaining: 9.66s
26:	learn: 0.2775374	total: 948ms	remaining: 9.58s
27:	learn: 0.2696916	total: 979ms	remaining: 9.51s
28:	learn: 0.2625452	total: 1.01s	remaining: 9.48s
29:	learn: 0.2558621	total: 1.04s	remaining: 9.41s
30:	learn: 0.2489345	total: 1.08s	remaining: 9.35s
31:	learn: 0.2421030	total: 1.11s	remaining: 9.29s
32:	learn: 0.2370184	total: 1.14s	remaining: 9.22s
33:	learn: 0.2324372	total: 1.17s	remaining: 9.16s
34:	learn: 0.2276736	total: 1.2s	remaining: 9.12s
35:	learn: 0.2216663	total: 1.24s	remaining: 9.06s
36:	learn: 0.2152706	total: 1.27s	remaining: 9.01s
37:	learn: 0.2112083	total: 1.3s	remaining: 8.95s
38:	learn: 0.2065521	total: 1.33s	remaining: 8.9s
39:	learn: 0.2005354	total: 1.36s	remaining: 8.85s
40:	learn: 0.1962722	total: 1.39s	

185:	learn: 0.0299320	total: 6.16s	remaining: 3.78s
186:	learn: 0.0297009	total: 6.19s	remaining: 3.74s
187:	learn: 0.0295415	total: 6.22s	remaining: 3.71s
188:	learn: 0.0293046	total: 6.25s	remaining: 3.67s
189:	learn: 0.0291545	total: 6.29s	remaining: 3.64s
190:	learn: 0.0290710	total: 6.32s	remaining: 3.6s
191:	learn: 0.0287610	total: 6.35s	remaining: 3.57s
192:	learn: 0.0285842	total: 6.38s	remaining: 3.54s
193:	learn: 0.0285541	total: 6.41s	remaining: 3.5s
194:	learn: 0.0283557	total: 6.44s	remaining: 3.47s
195:	learn: 0.0281262	total: 6.47s	remaining: 3.44s
196:	learn: 0.0278557	total: 6.5s	remaining: 3.4s
197:	learn: 0.0276103	total: 6.54s	remaining: 3.37s
198:	learn: 0.0273678	total: 6.57s	remaining: 3.33s
199:	learn: 0.0272775	total: 6.6s	remaining: 3.3s
200:	learn: 0.0270694	total: 6.63s	remaining: 3.27s
201:	learn: 0.0267701	total: 6.67s	remaining: 3.24s
202:	learn: 0.0264437	total: 6.71s	remaining: 3.21s
203:	learn: 0.0262161	total: 6.75s	remaining: 3.18s
204:	learn: 0.0259

49:	learn: 0.1567611	total: 1.62s	remaining: 8.11s
50:	learn: 0.1535128	total: 1.65s	remaining: 8.07s
51:	learn: 0.1505658	total: 1.68s	remaining: 8.03s
52:	learn: 0.1478219	total: 1.72s	remaining: 8s
53:	learn: 0.1446053	total: 1.75s	remaining: 7.96s
54:	learn: 0.1415665	total: 1.78s	remaining: 7.92s
55:	learn: 0.1390012	total: 1.81s	remaining: 7.9s
56:	learn: 0.1362393	total: 1.85s	remaining: 7.87s
57:	learn: 0.1340909	total: 1.88s	remaining: 7.83s
58:	learn: 0.1316157	total: 1.91s	remaining: 7.8s
59:	learn: 0.1286090	total: 1.94s	remaining: 7.76s
60:	learn: 0.1264025	total: 1.97s	remaining: 7.73s
61:	learn: 0.1241120	total: 2s	remaining: 7.69s
62:	learn: 0.1218843	total: 2.04s	remaining: 7.66s
63:	learn: 0.1205396	total: 2.07s	remaining: 7.62s
64:	learn: 0.1181439	total: 2.1s	remaining: 7.59s
65:	learn: 0.1156423	total: 2.13s	remaining: 7.55s
66:	learn: 0.1143575	total: 2.16s	remaining: 7.52s
67:	learn: 0.1119811	total: 2.19s	remaining: 7.48s
68:	learn: 0.1095692	total: 2.22s	remain

210:	learn: 0.0264865	total: 6.78s	remaining: 2.86s
211:	learn: 0.0262419	total: 6.81s	remaining: 2.83s
212:	learn: 0.0261065	total: 6.84s	remaining: 2.79s
213:	learn: 0.0258829	total: 6.88s	remaining: 2.76s
214:	learn: 0.0257557	total: 6.91s	remaining: 2.73s
215:	learn: 0.0256072	total: 6.94s	remaining: 2.7s
216:	learn: 0.0254083	total: 6.97s	remaining: 2.67s
217:	learn: 0.0252479	total: 7.01s	remaining: 2.63s
218:	learn: 0.0250279	total: 7.04s	remaining: 2.6s
219:	learn: 0.0248359	total: 7.07s	remaining: 2.57s
220:	learn: 0.0246279	total: 7.1s	remaining: 2.54s
221:	learn: 0.0246013	total: 7.13s	remaining: 2.51s
222:	learn: 0.0244587	total: 7.17s	remaining: 2.47s
223:	learn: 0.0243773	total: 7.2s	remaining: 2.44s
224:	learn: 0.0241800	total: 7.23s	remaining: 2.41s
225:	learn: 0.0240545	total: 7.26s	remaining: 2.38s
226:	learn: 0.0238253	total: 7.29s	remaining: 2.35s
227:	learn: 0.0236879	total: 7.33s	remaining: 2.31s
228:	learn: 0.0234678	total: 7.36s	remaining: 2.28s
229:	learn: 0.02

74:	learn: 0.1163167	total: 2.47s	remaining: 7.4s
75:	learn: 0.1141931	total: 2.5s	remaining: 7.37s
76:	learn: 0.1125071	total: 2.53s	remaining: 7.33s
77:	learn: 0.1111417	total: 2.56s	remaining: 7.29s
78:	learn: 0.1096003	total: 2.59s	remaining: 7.26s
79:	learn: 0.1080895	total: 2.63s	remaining: 7.22s
80:	learn: 0.1065784	total: 2.66s	remaining: 7.18s
81:	learn: 0.1051928	total: 2.69s	remaining: 7.15s
82:	learn: 0.1040770	total: 2.72s	remaining: 7.12s
83:	learn: 0.1027630	total: 2.75s	remaining: 7.08s
84:	learn: 0.1012706	total: 2.79s	remaining: 7.05s
85:	learn: 0.1001124	total: 2.82s	remaining: 7.01s
86:	learn: 0.0986794	total: 2.85s	remaining: 6.98s
87:	learn: 0.0973328	total: 2.88s	remaining: 6.94s
88:	learn: 0.0957423	total: 2.91s	remaining: 6.91s
89:	learn: 0.0945668	total: 2.95s	remaining: 6.87s
90:	learn: 0.0933797	total: 2.98s	remaining: 6.84s
91:	learn: 0.0925085	total: 3.01s	remaining: 6.8s
92:	learn: 0.0914085	total: 3.04s	remaining: 6.77s
93:	learn: 0.0901108	total: 3.07s	

237:	learn: 0.0249228	total: 7.8s	remaining: 2.03s
238:	learn: 0.0247583	total: 7.84s	remaining: 2s
239:	learn: 0.0245447	total: 7.87s	remaining: 1.97s
240:	learn: 0.0243358	total: 7.9s	remaining: 1.94s
241:	learn: 0.0241098	total: 7.94s	remaining: 1.9s
242:	learn: 0.0238925	total: 7.97s	remaining: 1.87s
243:	learn: 0.0237551	total: 8.01s	remaining: 1.84s
244:	learn: 0.0235605	total: 8.04s	remaining: 1.8s
245:	learn: 0.0234338	total: 8.08s	remaining: 1.77s
246:	learn: 0.0234075	total: 8.11s	remaining: 1.74s
247:	learn: 0.0232363	total: 8.14s	remaining: 1.71s
248:	learn: 0.0231101	total: 8.18s	remaining: 1.67s
249:	learn: 0.0229306	total: 8.21s	remaining: 1.64s
250:	learn: 0.0228270	total: 8.24s	remaining: 1.61s
251:	learn: 0.0226801	total: 8.28s	remaining: 1.58s
252:	learn: 0.0224358	total: 8.31s	remaining: 1.54s
253:	learn: 0.0224036	total: 8.34s	remaining: 1.51s
254:	learn: 0.0222937	total: 8.38s	remaining: 1.48s
255:	learn: 0.0221476	total: 8.41s	remaining: 1.45s
256:	learn: 0.02199

98:	learn: 0.1286537	total: 3.2s	remaining: 13s
99:	learn: 0.1273128	total: 3.25s	remaining: 13s
100:	learn: 0.1248584	total: 3.28s	remaining: 13s
101:	learn: 0.1238140	total: 3.31s	remaining: 12.9s
102:	learn: 0.1229108	total: 3.35s	remaining: 12.9s
103:	learn: 0.1215955	total: 3.38s	remaining: 12.9s
104:	learn: 0.1201878	total: 3.41s	remaining: 12.8s
105:	learn: 0.1190600	total: 3.44s	remaining: 12.8s
106:	learn: 0.1179818	total: 3.47s	remaining: 12.8s
107:	learn: 0.1169665	total: 3.5s	remaining: 12.7s
108:	learn: 0.1156193	total: 3.54s	remaining: 12.7s
109:	learn: 0.1147454	total: 3.57s	remaining: 12.7s
110:	learn: 0.1138533	total: 3.6s	remaining: 12.6s
111:	learn: 0.1125593	total: 3.63s	remaining: 12.6s
112:	learn: 0.1116370	total: 3.67s	remaining: 12.6s
113:	learn: 0.1103296	total: 3.7s	remaining: 12.5s
114:	learn: 0.1094820	total: 3.73s	remaining: 12.5s
115:	learn: 0.1079644	total: 3.76s	remaining: 12.4s
116:	learn: 0.1070956	total: 3.79s	remaining: 12.4s
117:	learn: 0.1061719	to

262:	learn: 0.0369309	total: 8.59s	remaining: 7.74s
263:	learn: 0.0366475	total: 8.63s	remaining: 7.71s
264:	learn: 0.0365581	total: 8.66s	remaining: 7.68s
265:	learn: 0.0363300	total: 8.69s	remaining: 7.64s
266:	learn: 0.0361463	total: 8.72s	remaining: 7.61s
267:	learn: 0.0358878	total: 8.75s	remaining: 7.58s
268:	learn: 0.0356287	total: 8.79s	remaining: 7.54s
269:	learn: 0.0354969	total: 8.82s	remaining: 7.51s
270:	learn: 0.0353614	total: 8.85s	remaining: 7.48s
271:	learn: 0.0352367	total: 8.88s	remaining: 7.45s
272:	learn: 0.0350356	total: 8.92s	remaining: 7.41s
273:	learn: 0.0347752	total: 8.95s	remaining: 7.38s
274:	learn: 0.0346057	total: 8.98s	remaining: 7.35s
275:	learn: 0.0343250	total: 9.01s	remaining: 7.31s
276:	learn: 0.0342116	total: 9.04s	remaining: 7.28s
277:	learn: 0.0340798	total: 9.08s	remaining: 7.25s
278:	learn: 0.0339728	total: 9.11s	remaining: 7.21s
279:	learn: 0.0338833	total: 9.14s	remaining: 7.18s
280:	learn: 0.0338279	total: 9.17s	remaining: 7.15s
281:	learn: 

423:	learn: 0.0170694	total: 13.8s	remaining: 2.47s
424:	learn: 0.0169935	total: 13.8s	remaining: 2.44s
425:	learn: 0.0169498	total: 13.8s	remaining: 2.4s
426:	learn: 0.0168635	total: 13.9s	remaining: 2.37s
427:	learn: 0.0168125	total: 13.9s	remaining: 2.34s
428:	learn: 0.0167086	total: 13.9s	remaining: 2.31s
429:	learn: 0.0166152	total: 14s	remaining: 2.27s
430:	learn: 0.0165017	total: 14s	remaining: 2.24s
431:	learn: 0.0164587	total: 14s	remaining: 2.21s
432:	learn: 0.0163805	total: 14.1s	remaining: 2.17s
433:	learn: 0.0163481	total: 14.1s	remaining: 2.14s
434:	learn: 0.0162832	total: 14.1s	remaining: 2.11s
435:	learn: 0.0162462	total: 14.2s	remaining: 2.08s
436:	learn: 0.0161495	total: 14.2s	remaining: 2.04s
437:	learn: 0.0160824	total: 14.2s	remaining: 2.01s
438:	learn: 0.0160261	total: 14.3s	remaining: 1.98s
439:	learn: 0.0159457	total: 14.3s	remaining: 1.95s
440:	learn: 0.0158522	total: 14.3s	remaining: 1.91s
441:	learn: 0.0158018	total: 14.3s	remaining: 1.88s
442:	learn: 0.01571

89:	learn: 0.1379479	total: 2.9s	remaining: 13.2s
90:	learn: 0.1368550	total: 2.94s	remaining: 13.2s
91:	learn: 0.1355753	total: 2.97s	remaining: 13.2s
92:	learn: 0.1342445	total: 3s	remaining: 13.1s
93:	learn: 0.1331347	total: 3.03s	remaining: 13.1s
94:	learn: 0.1320209	total: 3.06s	remaining: 13.1s
95:	learn: 0.1309129	total: 3.1s	remaining: 13s
96:	learn: 0.1295761	total: 3.13s	remaining: 13s
97:	learn: 0.1277998	total: 3.16s	remaining: 13s
98:	learn: 0.1266510	total: 3.19s	remaining: 12.9s
99:	learn: 0.1252084	total: 3.22s	remaining: 12.9s
100:	learn: 0.1238959	total: 3.25s	remaining: 12.9s
101:	learn: 0.1229094	total: 3.29s	remaining: 12.8s
102:	learn: 0.1213840	total: 3.32s	remaining: 12.8s
103:	learn: 0.1197898	total: 3.35s	remaining: 12.8s
104:	learn: 0.1186810	total: 3.38s	remaining: 12.7s
105:	learn: 0.1171420	total: 3.41s	remaining: 12.7s
106:	learn: 0.1157498	total: 3.44s	remaining: 12.7s
107:	learn: 0.1151845	total: 3.48s	remaining: 12.6s
108:	learn: 0.1141567	total: 3.51s

250:	learn: 0.0409335	total: 8.06s	remaining: 7.99s
251:	learn: 0.0406717	total: 8.09s	remaining: 7.96s
252:	learn: 0.0403964	total: 8.12s	remaining: 7.93s
253:	learn: 0.0400405	total: 8.15s	remaining: 7.89s
254:	learn: 0.0399619	total: 8.18s	remaining: 7.86s
255:	learn: 0.0396431	total: 8.21s	remaining: 7.83s
256:	learn: 0.0393133	total: 8.25s	remaining: 7.8s
257:	learn: 0.0390022	total: 8.28s	remaining: 7.76s
258:	learn: 0.0387135	total: 8.31s	remaining: 7.73s
259:	learn: 0.0384158	total: 8.34s	remaining: 7.7s
260:	learn: 0.0381494	total: 8.37s	remaining: 7.67s
261:	learn: 0.0378765	total: 8.4s	remaining: 7.63s
262:	learn: 0.0378148	total: 8.44s	remaining: 7.6s
263:	learn: 0.0375202	total: 8.47s	remaining: 7.57s
264:	learn: 0.0373113	total: 8.5s	remaining: 7.54s
265:	learn: 0.0369960	total: 8.53s	remaining: 7.51s
266:	learn: 0.0367698	total: 8.57s	remaining: 7.47s
267:	learn: 0.0365383	total: 8.6s	remaining: 7.44s
268:	learn: 0.0363552	total: 8.63s	remaining: 7.41s
269:	learn: 0.0360

411:	learn: 0.0189445	total: 13.2s	remaining: 2.82s
412:	learn: 0.0189028	total: 13.2s	remaining: 2.79s
413:	learn: 0.0188719	total: 13.3s	remaining: 2.75s
414:	learn: 0.0188409	total: 13.3s	remaining: 2.72s
415:	learn: 0.0187776	total: 13.3s	remaining: 2.69s
416:	learn: 0.0187063	total: 13.3s	remaining: 2.66s
417:	learn: 0.0186153	total: 13.4s	remaining: 2.62s
418:	learn: 0.0185486	total: 13.4s	remaining: 2.59s
419:	learn: 0.0184294	total: 13.4s	remaining: 2.56s
420:	learn: 0.0183231	total: 13.5s	remaining: 2.53s
421:	learn: 0.0182449	total: 13.5s	remaining: 2.5s
422:	learn: 0.0182327	total: 13.5s	remaining: 2.46s
423:	learn: 0.0181186	total: 13.6s	remaining: 2.43s
424:	learn: 0.0180150	total: 13.6s	remaining: 2.4s
425:	learn: 0.0180043	total: 13.6s	remaining: 2.37s
426:	learn: 0.0179843	total: 13.7s	remaining: 2.34s
427:	learn: 0.0179183	total: 13.7s	remaining: 2.3s
428:	learn: 0.0178259	total: 13.7s	remaining: 2.27s
429:	learn: 0.0177501	total: 13.8s	remaining: 2.24s
430:	learn: 0.0

72:	learn: 0.1860568	total: 2.46s	remaining: 14.4s
73:	learn: 0.1835412	total: 2.49s	remaining: 14.3s
74:	learn: 0.1812046	total: 2.52s	remaining: 14.3s
75:	learn: 0.1791167	total: 2.56s	remaining: 14.3s
76:	learn: 0.1775755	total: 2.59s	remaining: 14.2s
77:	learn: 0.1754466	total: 2.63s	remaining: 14.2s
78:	learn: 0.1735375	total: 2.66s	remaining: 14.2s
79:	learn: 0.1713510	total: 2.69s	remaining: 14.1s
80:	learn: 0.1699047	total: 2.73s	remaining: 14.1s
81:	learn: 0.1673550	total: 2.76s	remaining: 14.1s
82:	learn: 0.1655741	total: 2.8s	remaining: 14.1s
83:	learn: 0.1636839	total: 2.83s	remaining: 14s
84:	learn: 0.1616760	total: 2.86s	remaining: 14s
85:	learn: 0.1600636	total: 2.9s	remaining: 14s
86:	learn: 0.1582040	total: 2.94s	remaining: 13.9s
87:	learn: 0.1566444	total: 2.97s	remaining: 13.9s
88:	learn: 0.1547885	total: 3s	remaining: 13.9s
89:	learn: 0.1530620	total: 3.03s	remaining: 13.8s
90:	learn: 0.1518189	total: 3.07s	remaining: 13.8s
91:	learn: 0.1498724	total: 3.1s	remaining

234:	learn: 0.0489628	total: 7.82s	remaining: 8.82s
235:	learn: 0.0486776	total: 7.86s	remaining: 8.79s
236:	learn: 0.0483436	total: 7.89s	remaining: 8.75s
237:	learn: 0.0480522	total: 7.92s	remaining: 8.72s
238:	learn: 0.0477264	total: 7.96s	remaining: 8.69s
239:	learn: 0.0472295	total: 7.99s	remaining: 8.66s
240:	learn: 0.0470416	total: 8.02s	remaining: 8.62s
241:	learn: 0.0467501	total: 8.06s	remaining: 8.59s
242:	learn: 0.0464109	total: 8.09s	remaining: 8.56s
243:	learn: 0.0463715	total: 8.13s	remaining: 8.52s
244:	learn: 0.0461469	total: 8.16s	remaining: 8.49s
245:	learn: 0.0457645	total: 8.19s	remaining: 8.46s
246:	learn: 0.0454920	total: 8.22s	remaining: 8.42s
247:	learn: 0.0451977	total: 8.26s	remaining: 8.39s
248:	learn: 0.0449693	total: 8.29s	remaining: 8.36s
249:	learn: 0.0448919	total: 8.33s	remaining: 8.33s
250:	learn: 0.0446515	total: 8.36s	remaining: 8.29s
251:	learn: 0.0444629	total: 8.39s	remaining: 8.26s
252:	learn: 0.0441208	total: 8.43s	remaining: 8.23s
253:	learn: 

397:	learn: 0.0221894	total: 13.2s	remaining: 3.37s
398:	learn: 0.0221138	total: 13.2s	remaining: 3.34s
399:	learn: 0.0220913	total: 13.2s	remaining: 3.31s
400:	learn: 0.0220514	total: 13.3s	remaining: 3.27s
401:	learn: 0.0220345	total: 13.3s	remaining: 3.24s
402:	learn: 0.0219017	total: 13.3s	remaining: 3.21s
403:	learn: 0.0218130	total: 13.3s	remaining: 3.17s
404:	learn: 0.0217467	total: 13.4s	remaining: 3.14s
405:	learn: 0.0216248	total: 13.4s	remaining: 3.1s
406:	learn: 0.0215355	total: 13.4s	remaining: 3.07s
407:	learn: 0.0214001	total: 13.5s	remaining: 3.04s
408:	learn: 0.0213102	total: 13.5s	remaining: 3s
409:	learn: 0.0212216	total: 13.5s	remaining: 2.97s
410:	learn: 0.0210865	total: 13.6s	remaining: 2.94s
411:	learn: 0.0210012	total: 13.6s	remaining: 2.91s
412:	learn: 0.0209015	total: 13.6s	remaining: 2.87s
413:	learn: 0.0207852	total: 13.7s	remaining: 2.84s
414:	learn: 0.0206767	total: 13.7s	remaining: 2.81s
415:	learn: 0.0206673	total: 13.7s	remaining: 2.77s
416:	learn: 0.02

56:	learn: 0.0371459	total: 3.23s	remaining: 2.43s
57:	learn: 0.0360303	total: 3.28s	remaining: 2.38s
58:	learn: 0.0352459	total: 3.34s	remaining: 2.32s
59:	learn: 0.0342716	total: 3.39s	remaining: 2.26s
60:	learn: 0.0334574	total: 3.45s	remaining: 2.21s
61:	learn: 0.0326724	total: 3.51s	remaining: 2.15s
62:	learn: 0.0320990	total: 3.56s	remaining: 2.09s
63:	learn: 0.0315089	total: 3.62s	remaining: 2.03s
64:	learn: 0.0307268	total: 3.67s	remaining: 1.98s
65:	learn: 0.0303169	total: 3.73s	remaining: 1.92s
66:	learn: 0.0295871	total: 3.78s	remaining: 1.86s
67:	learn: 0.0290653	total: 3.84s	remaining: 1.8s
68:	learn: 0.0283633	total: 3.89s	remaining: 1.75s
69:	learn: 0.0276766	total: 3.95s	remaining: 1.69s
70:	learn: 0.0269039	total: 4s	remaining: 1.63s
71:	learn: 0.0261154	total: 4.06s	remaining: 1.58s
72:	learn: 0.0256204	total: 4.11s	remaining: 1.52s
73:	learn: 0.0253331	total: 4.17s	remaining: 1.46s
74:	learn: 0.0249333	total: 4.22s	remaining: 1.41s
75:	learn: 0.0244765	total: 4.28s	r

20:	learn: 0.1472788	total: 1.21s	remaining: 4.54s
21:	learn: 0.1404502	total: 1.26s	remaining: 4.47s
22:	learn: 0.1341615	total: 1.32s	remaining: 4.41s
23:	learn: 0.1289109	total: 1.37s	remaining: 4.34s
24:	learn: 0.1225951	total: 1.43s	remaining: 4.28s
25:	learn: 0.1165379	total: 1.49s	remaining: 4.23s
26:	learn: 0.1116958	total: 1.54s	remaining: 4.17s
27:	learn: 0.1081499	total: 1.6s	remaining: 4.11s
28:	learn: 0.1045452	total: 1.66s	remaining: 4.05s
29:	learn: 0.1005574	total: 1.71s	remaining: 4s
30:	learn: 0.0969839	total: 1.77s	remaining: 3.94s
31:	learn: 0.0936191	total: 1.82s	remaining: 3.88s
32:	learn: 0.0912838	total: 1.88s	remaining: 3.82s
33:	learn: 0.0879046	total: 1.94s	remaining: 3.76s
34:	learn: 0.0851626	total: 1.99s	remaining: 3.7s
35:	learn: 0.0818511	total: 2.05s	remaining: 3.64s
36:	learn: 0.0793190	total: 2.1s	remaining: 3.58s
37:	learn: 0.0762594	total: 2.16s	remaining: 3.52s
38:	learn: 0.0734129	total: 2.21s	remaining: 3.46s
39:	learn: 0.0713275	total: 2.27s	rem

82:	learn: 0.0899244	total: 4.91s	remaining: 12.8s
83:	learn: 0.0884181	total: 4.97s	remaining: 12.8s
84:	learn: 0.0866497	total: 5.02s	remaining: 12.7s
85:	learn: 0.0854661	total: 5.08s	remaining: 12.6s
86:	learn: 0.0842233	total: 5.13s	remaining: 12.6s
87:	learn: 0.0831204	total: 5.19s	remaining: 12.5s
88:	learn: 0.0817897	total: 5.25s	remaining: 12.4s
89:	learn: 0.0801934	total: 5.3s	remaining: 12.4s
90:	learn: 0.0788937	total: 5.36s	remaining: 12.3s
91:	learn: 0.0777876	total: 5.41s	remaining: 12.2s
92:	learn: 0.0765640	total: 5.47s	remaining: 12.2s
93:	learn: 0.0748926	total: 5.52s	remaining: 12.1s
94:	learn: 0.0738868	total: 5.58s	remaining: 12s
95:	learn: 0.0729607	total: 5.64s	remaining: 12s
96:	learn: 0.0718626	total: 5.7s	remaining: 11.9s
97:	learn: 0.0707111	total: 5.75s	remaining: 11.9s
98:	learn: 0.0700753	total: 5.81s	remaining: 11.8s
99:	learn: 0.0691687	total: 5.87s	remaining: 11.7s
100:	learn: 0.0681854	total: 5.92s	remaining: 11.7s
101:	learn: 0.0673318	total: 5.98s	r

242:	learn: 0.0180356	total: 13.9s	remaining: 3.27s
243:	learn: 0.0178588	total: 14s	remaining: 3.21s
244:	learn: 0.0177642	total: 14s	remaining: 3.15s
245:	learn: 0.0176269	total: 14.1s	remaining: 3.1s
246:	learn: 0.0174903	total: 14.2s	remaining: 3.04s
247:	learn: 0.0173992	total: 14.2s	remaining: 2.98s
248:	learn: 0.0172920	total: 14.3s	remaining: 2.92s
249:	learn: 0.0171905	total: 14.3s	remaining: 2.87s
250:	learn: 0.0170224	total: 14.4s	remaining: 2.81s
251:	learn: 0.0169136	total: 14.4s	remaining: 2.75s
252:	learn: 0.0167982	total: 14.5s	remaining: 2.69s
253:	learn: 0.0166939	total: 14.6s	remaining: 2.63s
254:	learn: 0.0165699	total: 14.6s	remaining: 2.58s
255:	learn: 0.0164743	total: 14.7s	remaining: 2.52s
256:	learn: 0.0163491	total: 14.7s	remaining: 2.46s
257:	learn: 0.0162800	total: 14.8s	remaining: 2.4s
258:	learn: 0.0161410	total: 14.8s	remaining: 2.35s
259:	learn: 0.0160796	total: 14.9s	remaining: 2.29s
260:	learn: 0.0159556	total: 14.9s	remaining: 2.23s
261:	learn: 0.0158

104:	learn: 0.0641231	total: 6.04s	remaining: 11.2s
105:	learn: 0.0632292	total: 6.1s	remaining: 11.2s
106:	learn: 0.0624213	total: 6.16s	remaining: 11.1s
107:	learn: 0.0617328	total: 6.21s	remaining: 11s
108:	learn: 0.0609751	total: 6.27s	remaining: 11s
109:	learn: 0.0602370	total: 6.32s	remaining: 10.9s
110:	learn: 0.0594489	total: 6.38s	remaining: 10.9s
111:	learn: 0.0589280	total: 6.43s	remaining: 10.8s
112:	learn: 0.0582608	total: 6.49s	remaining: 10.7s
113:	learn: 0.0576057	total: 6.54s	remaining: 10.7s
114:	learn: 0.0567032	total: 6.6s	remaining: 10.6s
115:	learn: 0.0556620	total: 6.65s	remaining: 10.6s
116:	learn: 0.0551111	total: 6.71s	remaining: 10.5s
117:	learn: 0.0544475	total: 6.76s	remaining: 10.4s
118:	learn: 0.0537586	total: 6.82s	remaining: 10.4s
119:	learn: 0.0530545	total: 6.87s	remaining: 10.3s
120:	learn: 0.0522497	total: 6.93s	remaining: 10.2s
121:	learn: 0.0515647	total: 6.99s	remaining: 10.2s
122:	learn: 0.0510371	total: 7.04s	remaining: 10.1s
123:	learn: 0.0504

264:	learn: 0.0155362	total: 15.2s	remaining: 2s
265:	learn: 0.0154460	total: 15.2s	remaining: 1.95s
266:	learn: 0.0153329	total: 15.3s	remaining: 1.89s
267:	learn: 0.0152335	total: 15.3s	remaining: 1.83s
268:	learn: 0.0151520	total: 15.4s	remaining: 1.77s
269:	learn: 0.0150611	total: 15.5s	remaining: 1.72s
270:	learn: 0.0149730	total: 15.5s	remaining: 1.66s
271:	learn: 0.0148724	total: 15.6s	remaining: 1.6s
272:	learn: 0.0147700	total: 15.6s	remaining: 1.54s
273:	learn: 0.0146886	total: 15.7s	remaining: 1.49s
274:	learn: 0.0146252	total: 15.7s	remaining: 1.43s
275:	learn: 0.0144982	total: 15.8s	remaining: 1.37s
276:	learn: 0.0144192	total: 15.8s	remaining: 1.31s
277:	learn: 0.0143241	total: 15.9s	remaining: 1.26s
278:	learn: 0.0142277	total: 16s	remaining: 1.2s
279:	learn: 0.0141109	total: 16s	remaining: 1.14s
280:	learn: 0.0140187	total: 16.1s	remaining: 1.09s
281:	learn: 0.0139337	total: 16.1s	remaining: 1.03s
282:	learn: 0.0138714	total: 16.2s	remaining: 972ms
283:	learn: 0.0137592

126:	learn: 0.0542423	total: 7.54s	remaining: 10.3s
127:	learn: 0.0537663	total: 7.6s	remaining: 10.2s
128:	learn: 0.0530226	total: 7.66s	remaining: 10.2s
129:	learn: 0.0526426	total: 7.71s	remaining: 10.1s
130:	learn: 0.0521977	total: 7.77s	remaining: 10s
131:	learn: 0.0516696	total: 7.83s	remaining: 9.96s
132:	learn: 0.0510430	total: 7.88s	remaining: 9.9s
133:	learn: 0.0505125	total: 7.94s	remaining: 9.84s
134:	learn: 0.0500104	total: 8.01s	remaining: 9.79s
135:	learn: 0.0494757	total: 8.08s	remaining: 9.74s
136:	learn: 0.0490509	total: 8.14s	remaining: 9.69s
137:	learn: 0.0486623	total: 8.21s	remaining: 9.64s
138:	learn: 0.0481381	total: 8.28s	remaining: 9.59s
139:	learn: 0.0476254	total: 8.34s	remaining: 9.53s
140:	learn: 0.0473854	total: 8.4s	remaining: 9.47s
141:	learn: 0.0470601	total: 8.45s	remaining: 9.41s
142:	learn: 0.0464418	total: 8.51s	remaining: 9.35s
143:	learn: 0.0459713	total: 8.57s	remaining: 9.28s
144:	learn: 0.0455294	total: 8.63s	remaining: 9.22s
145:	learn: 0.045

286:	learn: 0.0149001	total: 16.6s	remaining: 751ms
287:	learn: 0.0147841	total: 16.6s	remaining: 693ms
288:	learn: 0.0146557	total: 16.7s	remaining: 636ms
289:	learn: 0.0145407	total: 16.8s	remaining: 578ms
290:	learn: 0.0144290	total: 16.8s	remaining: 520ms
291:	learn: 0.0143205	total: 16.9s	remaining: 462ms
292:	learn: 0.0142250	total: 16.9s	remaining: 404ms
293:	learn: 0.0141511	total: 17s	remaining: 346ms
294:	learn: 0.0140655	total: 17s	remaining: 289ms
295:	learn: 0.0140306	total: 17.1s	remaining: 231ms
296:	learn: 0.0139268	total: 17.1s	remaining: 173ms
297:	learn: 0.0138144	total: 17.2s	remaining: 115ms
298:	learn: 0.0137442	total: 17.3s	remaining: 57.7ms
299:	learn: 0.0136718	total: 17.3s	remaining: 0us
Learning rate set to 0.014469
0:	learn: 0.6750337	total: 98.9ms	remaining: 49.3s
1:	learn: 0.6568555	total: 152ms	remaining: 37.8s
2:	learn: 0.6415315	total: 206ms	remaining: 34.1s
3:	learn: 0.6270140	total: 259ms	remaining: 32.2s
4:	learn: 0.6135074	total: 315ms	remaining: 31

148:	learn: 0.0765694	total: 8.39s	remaining: 19.8s
149:	learn: 0.0757607	total: 8.44s	remaining: 19.7s
150:	learn: 0.0750887	total: 8.5s	remaining: 19.6s
151:	learn: 0.0743378	total: 8.55s	remaining: 19.6s
152:	learn: 0.0737231	total: 8.61s	remaining: 19.5s
153:	learn: 0.0729183	total: 8.66s	remaining: 19.5s
154:	learn: 0.0722317	total: 8.72s	remaining: 19.4s
155:	learn: 0.0716527	total: 8.77s	remaining: 19.3s
156:	learn: 0.0708614	total: 8.83s	remaining: 19.3s
157:	learn: 0.0701950	total: 8.88s	remaining: 19.2s
158:	learn: 0.0697427	total: 8.94s	remaining: 19.2s
159:	learn: 0.0690566	total: 8.99s	remaining: 19.1s
160:	learn: 0.0684651	total: 9.04s	remaining: 19s
161:	learn: 0.0679079	total: 9.1s	remaining: 19s
162:	learn: 0.0672591	total: 9.16s	remaining: 18.9s
163:	learn: 0.0667051	total: 9.21s	remaining: 18.9s
164:	learn: 0.0660858	total: 9.27s	remaining: 18.8s
165:	learn: 0.0656239	total: 9.32s	remaining: 18.8s
166:	learn: 0.0649649	total: 9.38s	remaining: 18.7s
167:	learn: 0.0644

308:	learn: 0.0259492	total: 17.4s	remaining: 10.7s
309:	learn: 0.0257903	total: 17.4s	remaining: 10.7s
310:	learn: 0.0256928	total: 17.5s	remaining: 10.6s
311:	learn: 0.0254923	total: 17.5s	remaining: 10.6s
312:	learn: 0.0253497	total: 17.6s	remaining: 10.5s
313:	learn: 0.0252150	total: 17.6s	remaining: 10.5s
314:	learn: 0.0250724	total: 17.7s	remaining: 10.4s
315:	learn: 0.0249500	total: 17.8s	remaining: 10.3s
316:	learn: 0.0248506	total: 17.8s	remaining: 10.3s
317:	learn: 0.0247203	total: 17.9s	remaining: 10.2s
318:	learn: 0.0246182	total: 17.9s	remaining: 10.2s
319:	learn: 0.0244838	total: 18s	remaining: 10.1s
320:	learn: 0.0243478	total: 18s	remaining: 10.1s
321:	learn: 0.0242292	total: 18.1s	remaining: 10s
322:	learn: 0.0240866	total: 18.1s	remaining: 9.95s
323:	learn: 0.0239881	total: 18.2s	remaining: 9.89s
324:	learn: 0.0238460	total: 18.3s	remaining: 9.83s
325:	learn: 0.0236949	total: 18.3s	remaining: 9.78s
326:	learn: 0.0235605	total: 18.4s	remaining: 9.72s
327:	learn: 0.0234

468:	learn: 0.0127059	total: 26.3s	remaining: 1.74s
469:	learn: 0.0126573	total: 26.3s	remaining: 1.68s
470:	learn: 0.0126150	total: 26.4s	remaining: 1.62s
471:	learn: 0.0125502	total: 26.4s	remaining: 1.57s
472:	learn: 0.0124968	total: 26.5s	remaining: 1.51s
473:	learn: 0.0124436	total: 26.6s	remaining: 1.46s
474:	learn: 0.0123882	total: 26.6s	remaining: 1.4s
475:	learn: 0.0123328	total: 26.7s	remaining: 1.34s
476:	learn: 0.0122936	total: 26.7s	remaining: 1.29s
477:	learn: 0.0122257	total: 26.8s	remaining: 1.23s
478:	learn: 0.0121950	total: 26.8s	remaining: 1.18s
479:	learn: 0.0121422	total: 26.9s	remaining: 1.12s
480:	learn: 0.0120917	total: 26.9s	remaining: 1.06s
481:	learn: 0.0120468	total: 27s	remaining: 1.01s
482:	learn: 0.0119983	total: 27.1s	remaining: 952ms
483:	learn: 0.0119349	total: 27.1s	remaining: 896ms
484:	learn: 0.0118835	total: 27.2s	remaining: 840ms
485:	learn: 0.0118339	total: 27.2s	remaining: 784ms
486:	learn: 0.0117635	total: 27.3s	remaining: 728ms
487:	learn: 0.0

130:	learn: 0.0885757	total: 7.53s	remaining: 21.2s
131:	learn: 0.0877500	total: 7.59s	remaining: 21.2s
132:	learn: 0.0870616	total: 7.65s	remaining: 21.1s
133:	learn: 0.0865178	total: 7.7s	remaining: 21s
134:	learn: 0.0857726	total: 7.76s	remaining: 21s
135:	learn: 0.0849813	total: 7.82s	remaining: 20.9s
136:	learn: 0.0844309	total: 7.87s	remaining: 20.9s
137:	learn: 0.0837719	total: 7.93s	remaining: 20.8s
138:	learn: 0.0829325	total: 7.99s	remaining: 20.7s
139:	learn: 0.0821135	total: 8.04s	remaining: 20.7s
140:	learn: 0.0816423	total: 8.1s	remaining: 20.6s
141:	learn: 0.0806366	total: 8.15s	remaining: 20.6s
142:	learn: 0.0797724	total: 8.21s	remaining: 20.5s
143:	learn: 0.0790904	total: 8.27s	remaining: 20.4s
144:	learn: 0.0782728	total: 8.32s	remaining: 20.4s
145:	learn: 0.0776401	total: 8.38s	remaining: 20.3s
146:	learn: 0.0769679	total: 8.45s	remaining: 20.3s
147:	learn: 0.0762726	total: 8.5s	remaining: 20.2s
148:	learn: 0.0756175	total: 8.56s	remaining: 20.2s
149:	learn: 0.07494

290:	learn: 0.0288255	total: 17s	remaining: 12.2s
291:	learn: 0.0286888	total: 17.1s	remaining: 12.2s
292:	learn: 0.0285519	total: 17.1s	remaining: 12.1s
293:	learn: 0.0284203	total: 17.2s	remaining: 12.1s
294:	learn: 0.0282098	total: 17.3s	remaining: 12s
295:	learn: 0.0281016	total: 17.3s	remaining: 11.9s
296:	learn: 0.0279728	total: 17.4s	remaining: 11.9s
297:	learn: 0.0278469	total: 17.4s	remaining: 11.8s
298:	learn: 0.0277302	total: 17.5s	remaining: 11.8s
299:	learn: 0.0275865	total: 17.5s	remaining: 11.7s
300:	learn: 0.0274632	total: 17.6s	remaining: 11.6s
301:	learn: 0.0273522	total: 17.6s	remaining: 11.6s
302:	learn: 0.0271972	total: 17.7s	remaining: 11.5s
303:	learn: 0.0270520	total: 17.8s	remaining: 11.5s
304:	learn: 0.0269394	total: 17.8s	remaining: 11.4s
305:	learn: 0.0267560	total: 17.9s	remaining: 11.3s
306:	learn: 0.0266239	total: 17.9s	remaining: 11.3s
307:	learn: 0.0265452	total: 18s	remaining: 11.2s
308:	learn: 0.0264066	total: 18s	remaining: 11.2s
309:	learn: 0.026264

450:	learn: 0.0141017	total: 26.4s	remaining: 2.87s
451:	learn: 0.0140469	total: 26.5s	remaining: 2.81s
452:	learn: 0.0139946	total: 26.6s	remaining: 2.75s
453:	learn: 0.0139405	total: 26.6s	remaining: 2.69s
454:	learn: 0.0138809	total: 26.7s	remaining: 2.64s
455:	learn: 0.0138213	total: 26.7s	remaining: 2.58s
456:	learn: 0.0137694	total: 26.8s	remaining: 2.52s
457:	learn: 0.0137184	total: 26.9s	remaining: 2.46s
458:	learn: 0.0136594	total: 26.9s	remaining: 2.4s
459:	learn: 0.0136078	total: 27s	remaining: 2.35s
460:	learn: 0.0135452	total: 27s	remaining: 2.29s
461:	learn: 0.0134793	total: 27.1s	remaining: 2.23s
462:	learn: 0.0134145	total: 27.1s	remaining: 2.17s
463:	learn: 0.0133452	total: 27.2s	remaining: 2.11s
464:	learn: 0.0132817	total: 27.3s	remaining: 2.05s
465:	learn: 0.0132239	total: 27.3s	remaining: 1.99s
466:	learn: 0.0131798	total: 27.4s	remaining: 1.93s
467:	learn: 0.0130988	total: 27.4s	remaining: 1.88s
468:	learn: 0.0130485	total: 27.5s	remaining: 1.82s
469:	learn: 0.013

112:	learn: 0.1159202	total: 6.55s	remaining: 22.4s
113:	learn: 0.1146951	total: 6.61s	remaining: 22.4s
114:	learn: 0.1136019	total: 6.67s	remaining: 22.3s
115:	learn: 0.1122682	total: 6.72s	remaining: 22.3s
116:	learn: 0.1110683	total: 6.78s	remaining: 22.2s
117:	learn: 0.1100533	total: 6.83s	remaining: 22.1s
118:	learn: 0.1091968	total: 6.89s	remaining: 22.1s
119:	learn: 0.1080367	total: 6.95s	remaining: 22s
120:	learn: 0.1070879	total: 7s	remaining: 21.9s
121:	learn: 0.1058952	total: 7.06s	remaining: 21.9s
122:	learn: 0.1047979	total: 7.12s	remaining: 21.8s
123:	learn: 0.1036144	total: 7.17s	remaining: 21.8s
124:	learn: 0.1027506	total: 7.23s	remaining: 21.7s
125:	learn: 0.1020672	total: 7.29s	remaining: 21.6s
126:	learn: 0.1009923	total: 7.34s	remaining: 21.6s
127:	learn: 0.1000244	total: 7.4s	remaining: 21.5s
128:	learn: 0.0988699	total: 7.46s	remaining: 21.4s
129:	learn: 0.0980161	total: 7.52s	remaining: 21.4s
130:	learn: 0.0971941	total: 7.58s	remaining: 21.4s
131:	learn: 0.0964

272:	learn: 0.0350920	total: 15.9s	remaining: 13.2s
273:	learn: 0.0349335	total: 15.9s	remaining: 13.1s
274:	learn: 0.0347338	total: 16s	remaining: 13.1s
275:	learn: 0.0346150	total: 16s	remaining: 13s
276:	learn: 0.0343895	total: 16.1s	remaining: 13s
277:	learn: 0.0341488	total: 16.2s	remaining: 12.9s
278:	learn: 0.0339155	total: 16.2s	remaining: 12.9s
279:	learn: 0.0337038	total: 16.3s	remaining: 12.8s
280:	learn: 0.0335257	total: 16.3s	remaining: 12.7s
281:	learn: 0.0333321	total: 16.4s	remaining: 12.7s
282:	learn: 0.0331547	total: 16.5s	remaining: 12.6s
283:	learn: 0.0330574	total: 16.5s	remaining: 12.6s
284:	learn: 0.0329341	total: 16.6s	remaining: 12.5s
285:	learn: 0.0327351	total: 16.6s	remaining: 12.4s
286:	learn: 0.0325731	total: 16.7s	remaining: 12.4s
287:	learn: 0.0323631	total: 16.7s	remaining: 12.3s
288:	learn: 0.0322125	total: 16.8s	remaining: 12.3s
289:	learn: 0.0321042	total: 16.9s	remaining: 12.2s
290:	learn: 0.0319751	total: 16.9s	remaining: 12.1s
291:	learn: 0.031762

432:	learn: 0.0170098	total: 25.1s	remaining: 3.88s
433:	learn: 0.0169289	total: 25.1s	remaining: 3.82s
434:	learn: 0.0168575	total: 25.2s	remaining: 3.76s
435:	learn: 0.0167623	total: 25.2s	remaining: 3.7s
436:	learn: 0.0167134	total: 25.3s	remaining: 3.64s
437:	learn: 0.0166437	total: 25.3s	remaining: 3.59s
438:	learn: 0.0165739	total: 25.4s	remaining: 3.53s
439:	learn: 0.0165381	total: 25.4s	remaining: 3.47s
440:	learn: 0.0164443	total: 25.5s	remaining: 3.41s
441:	learn: 0.0163649	total: 25.6s	remaining: 3.35s
442:	learn: 0.0162826	total: 25.6s	remaining: 3.3s
443:	learn: 0.0162222	total: 25.7s	remaining: 3.24s
444:	learn: 0.0161652	total: 25.8s	remaining: 3.18s
445:	learn: 0.0160916	total: 25.8s	remaining: 3.13s
446:	learn: 0.0160180	total: 25.9s	remaining: 3.07s
447:	learn: 0.0159647	total: 25.9s	remaining: 3.01s
448:	learn: 0.0158945	total: 26s	remaining: 2.95s
449:	learn: 0.0158028	total: 26.1s	remaining: 2.89s
450:	learn: 0.0157379	total: 26.1s	remaining: 2.84s
451:	learn: 0.01

103:	learn: 0.1360094	total: 1.37s	remaining: 5.21s
104:	learn: 0.1349715	total: 1.39s	remaining: 5.21s
105:	learn: 0.1339988	total: 1.4s	remaining: 5.2s
106:	learn: 0.1333332	total: 1.41s	remaining: 5.19s
107:	learn: 0.1323065	total: 1.42s	remaining: 5.17s
108:	learn: 0.1314626	total: 1.44s	remaining: 5.16s
109:	learn: 0.1307703	total: 1.45s	remaining: 5.14s
110:	learn: 0.1299123	total: 1.46s	remaining: 5.13s
111:	learn: 0.1290791	total: 1.48s	remaining: 5.12s
112:	learn: 0.1281853	total: 1.49s	remaining: 5.1s
113:	learn: 0.1271441	total: 1.5s	remaining: 5.09s
114:	learn: 0.1260938	total: 1.51s	remaining: 5.07s
115:	learn: 0.1257861	total: 1.53s	remaining: 5.06s
116:	learn: 0.1253879	total: 1.54s	remaining: 5.04s
117:	learn: 0.1242468	total: 1.55s	remaining: 5.03s
118:	learn: 0.1237048	total: 1.57s	remaining: 5.02s
119:	learn: 0.1228592	total: 1.58s	remaining: 5s
120:	learn: 0.1221057	total: 1.59s	remaining: 5s
121:	learn: 0.1216225	total: 1.61s	remaining: 4.98s
122:	learn: 0.1210417	

263:	learn: 0.0680517	total: 3.44s	remaining: 3.08s
264:	learn: 0.0676277	total: 3.46s	remaining: 3.06s
265:	learn: 0.0671365	total: 3.47s	remaining: 3.05s
266:	learn: 0.0670645	total: 3.48s	remaining: 3.04s
267:	learn: 0.0667662	total: 3.5s	remaining: 3.03s
268:	learn: 0.0666453	total: 3.51s	remaining: 3.01s
269:	learn: 0.0663948	total: 3.52s	remaining: 3s
270:	learn: 0.0660256	total: 3.53s	remaining: 2.99s
271:	learn: 0.0658642	total: 3.55s	remaining: 2.97s
272:	learn: 0.0656346	total: 3.56s	remaining: 2.96s
273:	learn: 0.0653682	total: 3.57s	remaining: 2.95s
274:	learn: 0.0651303	total: 3.58s	remaining: 2.93s
275:	learn: 0.0646332	total: 3.6s	remaining: 2.92s
276:	learn: 0.0643165	total: 3.61s	remaining: 2.91s
277:	learn: 0.0641442	total: 3.62s	remaining: 2.89s
278:	learn: 0.0640111	total: 3.64s	remaining: 2.88s
279:	learn: 0.0635618	total: 3.65s	remaining: 2.87s
280:	learn: 0.0631927	total: 3.66s	remaining: 2.85s
281:	learn: 0.0628880	total: 3.68s	remaining: 2.84s
282:	learn: 0.062

423:	learn: 0.0385391	total: 5.51s	remaining: 988ms
424:	learn: 0.0383888	total: 5.53s	remaining: 975ms
425:	learn: 0.0382476	total: 5.54s	remaining: 962ms
426:	learn: 0.0382289	total: 5.55s	remaining: 949ms
427:	learn: 0.0381797	total: 5.56s	remaining: 936ms
428:	learn: 0.0381043	total: 5.58s	remaining: 923ms
429:	learn: 0.0378708	total: 5.59s	remaining: 910ms
430:	learn: 0.0377426	total: 5.6s	remaining: 897ms
431:	learn: 0.0375882	total: 5.62s	remaining: 884ms
432:	learn: 0.0373075	total: 5.63s	remaining: 871ms
433:	learn: 0.0372680	total: 5.64s	remaining: 858ms
434:	learn: 0.0372534	total: 5.66s	remaining: 845ms
435:	learn: 0.0370838	total: 5.67s	remaining: 832ms
436:	learn: 0.0369591	total: 5.68s	remaining: 819ms
437:	learn: 0.0367631	total: 5.69s	remaining: 806ms
438:	learn: 0.0366515	total: 5.71s	remaining: 793ms
439:	learn: 0.0366265	total: 5.72s	remaining: 780ms
440:	learn: 0.0364158	total: 5.74s	remaining: 767ms
441:	learn: 0.0363206	total: 5.75s	remaining: 754ms
442:	learn: 0

In [22]:
print('Лучшее значение F1-меры в CatBoost:', result_cat.best_score_)
print('Лучшие параметры CatBoost:', result_cat.best_params_)

Лучшее значение F1-меры в CatBoost: 0.4985682597297842
Лучшие параметры CatBoost: {'depth': 4, 'iterations': 500}


In [23]:
# Проверим модель LightGBM на тестовой выборке
best_model_cat = search_cat.best_estimator_
best_model_cat.fit(features_train, target_train)
predictions_cat=best_model_cat.predict(features_test)

Learning rate set to 0.017204
0:	learn: 0.6747612	total: 12.8ms	remaining: 6.36s
1:	learn: 0.6586915	total: 24.9ms	remaining: 6.21s
2:	learn: 0.6394685	total: 37.2ms	remaining: 6.16s
3:	learn: 0.6229377	total: 49.7ms	remaining: 6.16s
4:	learn: 0.6021268	total: 63ms	remaining: 6.23s
5:	learn: 0.5857017	total: 76.1ms	remaining: 6.27s
6:	learn: 0.5700984	total: 88.7ms	remaining: 6.25s
7:	learn: 0.5534114	total: 101ms	remaining: 6.24s
8:	learn: 0.5413213	total: 114ms	remaining: 6.24s
9:	learn: 0.5283144	total: 127ms	remaining: 6.21s
10:	learn: 0.5133013	total: 139ms	remaining: 6.2s
11:	learn: 0.5036730	total: 152ms	remaining: 6.18s
12:	learn: 0.4929720	total: 165ms	remaining: 6.18s
13:	learn: 0.4801533	total: 180ms	remaining: 6.23s
14:	learn: 0.4666687	total: 192ms	remaining: 6.21s
15:	learn: 0.4563282	total: 208ms	remaining: 6.28s
16:	learn: 0.4438585	total: 221ms	remaining: 6.27s
17:	learn: 0.4310014	total: 234ms	remaining: 6.28s
18:	learn: 0.4215554	total: 247ms	remaining: 6.25s
19:	lea

171:	learn: 0.0965922	total: 2.25s	remaining: 4.29s
172:	learn: 0.0961181	total: 2.26s	remaining: 4.28s
173:	learn: 0.0955388	total: 2.28s	remaining: 4.27s
174:	learn: 0.0948129	total: 2.29s	remaining: 4.25s
175:	learn: 0.0944628	total: 2.3s	remaining: 4.24s
176:	learn: 0.0942092	total: 2.32s	remaining: 4.23s
177:	learn: 0.0939037	total: 2.33s	remaining: 4.21s
178:	learn: 0.0932346	total: 2.34s	remaining: 4.2s
179:	learn: 0.0926536	total: 2.35s	remaining: 4.18s
180:	learn: 0.0922282	total: 2.37s	remaining: 4.17s
181:	learn: 0.0918499	total: 2.38s	remaining: 4.16s
182:	learn: 0.0916631	total: 2.39s	remaining: 4.14s
183:	learn: 0.0915019	total: 2.4s	remaining: 4.13s
184:	learn: 0.0914224	total: 2.42s	remaining: 4.12s
185:	learn: 0.0911457	total: 2.43s	remaining: 4.1s
186:	learn: 0.0906068	total: 2.44s	remaining: 4.09s
187:	learn: 0.0899350	total: 2.46s	remaining: 4.08s
188:	learn: 0.0893878	total: 2.47s	remaining: 4.07s
189:	learn: 0.0888940	total: 2.48s	remaining: 4.05s
190:	learn: 0.08

331:	learn: 0.0531668	total: 4.33s	remaining: 2.19s
332:	learn: 0.0528212	total: 4.34s	remaining: 2.18s
333:	learn: 0.0526526	total: 4.35s	remaining: 2.16s
334:	learn: 0.0523236	total: 4.37s	remaining: 2.15s
335:	learn: 0.0519303	total: 4.38s	remaining: 2.14s
336:	learn: 0.0516968	total: 4.39s	remaining: 2.13s
337:	learn: 0.0515214	total: 4.41s	remaining: 2.11s
338:	learn: 0.0513378	total: 4.42s	remaining: 2.1s
339:	learn: 0.0512435	total: 4.43s	remaining: 2.09s
340:	learn: 0.0509456	total: 4.45s	remaining: 2.07s
341:	learn: 0.0508716	total: 4.46s	remaining: 2.06s
342:	learn: 0.0507020	total: 4.47s	remaining: 2.05s
343:	learn: 0.0504861	total: 4.49s	remaining: 2.03s
344:	learn: 0.0503233	total: 4.5s	remaining: 2.02s
345:	learn: 0.0502639	total: 4.51s	remaining: 2.01s
346:	learn: 0.0499944	total: 4.52s	remaining: 1.99s
347:	learn: 0.0499729	total: 4.54s	remaining: 1.98s
348:	learn: 0.0497724	total: 4.55s	remaining: 1.97s
349:	learn: 0.0495493	total: 4.56s	remaining: 1.96s
350:	learn: 0.

491:	learn: 0.0307759	total: 6.41s	remaining: 104ms
492:	learn: 0.0306122	total: 6.42s	remaining: 91.2ms
493:	learn: 0.0305947	total: 6.44s	remaining: 78.2ms
494:	learn: 0.0304124	total: 6.45s	remaining: 65.2ms
495:	learn: 0.0302942	total: 6.46s	remaining: 52.1ms
496:	learn: 0.0302515	total: 6.48s	remaining: 39.1ms
497:	learn: 0.0302373	total: 6.49s	remaining: 26.1ms
498:	learn: 0.0302203	total: 6.5s	remaining: 13ms
499:	learn: 0.0300245	total: 6.52s	remaining: 0us


In [24]:
print("F1-мера лучшей модели на тестовой выборке:", f1_score(target_test, predictions_cat))

F1-мера лучшей модели на тестовой выборке: 0.5714285714285715


## Вывод по BERT: 
Работает долго, поэтому построил модели только на малой части данных (1000 случайных строк). Удивительно, но на логистической регресси каким то образом получилась f1-мера > 0.75=)

# Модель без участия BERT

## Подготовка данных для классификации текстов

In [2]:
# Загрузим данные
df_wikishop = pd.read_csv('df_toxic.csv')
corpus = list(df_wikishop['text'].values.astype('U'))

In [3]:
# Напишем функцию для лемматизации текста
nltk.download('wordnet')
nltk.download('punkt')
lemmatizer = WordNetLemmatizer()
def lemm(text):   
    word_list = nltk.word_tokenize(text)
    return' '.join([lemmatizer.lemmatize(w) for w in word_list])

[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/aleksandrverlan/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     /Users/aleksandrverlan/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [4]:
# Напишем функцию для очистки текста
def clear_text(text):
    return ' '.join(re.sub(r'[^a-zA-Z]', ' ', text).split())

In [5]:
# Применим функции ко всем строкам
df_wikishop['text_lemm'] = df_wikishop['text'].apply(lambda x: lemm(clear_text(x)))

In [6]:
df_wikishop.head()

text  toxic  \
0  Explanation\nWhy the edits made under my usern...      0   
1  D'aww! He matches this background colour I'm s...      0   
2  Hey man, I'm really not trying to edit war. It...      0   
3  "\nMore\nI can't make any real suggestions on ...      0   
4  You, sir, are my hero. Any chance you remember...      0   

                                           text_lemm  
0  Explanation Why the edits made under my userna...  
1  D aww He match this background colour I m seem...  
2  Hey man I m really not trying to edit war It s...  
3  More I can t make any real suggestion on impro...  
4  You sir are my hero Any chance you remember wh...

In [7]:
df_wikishop.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 3 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   text       159571 non-null  object
 1   toxic      159571 non-null  int64 
 2   text_lemm  159571 non-null  object
dtypes: int64(1), object(2)
memory usage: 3.7+ MB


In [8]:
# Разделим данные на тренировочную и тестовую выборки
train, test = train_test_split(df_wikishop, test_size=0.25, random_state=12345)

In [12]:
nltk.download('stopwords')
stopwords = set(nltk_stopwords.words('english'))
# Выделим переменные признаки и признак, который нужно предсказать
train_target=train['toxic']
train_corpus=train['text_lemm']
count_tf_idf = TfidfVectorizer(stop_words=stopwords)
train_features = count_tf_idf.fit_transform(train_corpus) 
test_target=test['toxic']
test_corpus=test['text_lemm']
test_features = count_tf_idf.transform(test_corpus) 

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/aleksandrverlan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


## Построение моделей

In [13]:
# Создадим функцию обучения модели
def fit_predict(model):
    model.fit(train_features, train_target)
    predict=model.predict(test_features)
    f1=f1_score(test_target, predict)
    print("F1 модели на тестовой выборке:", f1)

### Модель логистической регрессии

In [14]:
print('Логистическая регрессия')
model_lr = LogisticRegression(random_state=12345)
fit_predict(model_lr)

Логистическая регрессия
F1 модели на тестовой выборке: 0.7374153664998528


### Модель LigthGBM

In [18]:
print('LightGBM')
model_lgb=LGBMClassifier(random_state=12345, n_estimators=500)
fit_predict(model_lgb)

LightGBM
F1 модели на тестовой выборке: 0.774986346258875


### Модель CatBoost

In [19]:
print('CatBoost')
model_cat=CatBoostClassifier(random_seed=12345, iterations=500)
fit_predict(model_cat)

CatBoost
Learning rate set to 0.150069
0:	learn: 0.5467665	total: 977ms	remaining: 8m 7s
1:	learn: 0.4518392	total: 1.71s	remaining: 7m 6s
2:	learn: 0.3855012	total: 2.18s	remaining: 6m 1s
3:	learn: 0.3389108	total: 2.65s	remaining: 5m 28s
4:	learn: 0.3079332	total: 3.12s	remaining: 5m 8s
5:	learn: 0.2842842	total: 3.64s	remaining: 4m 59s
6:	learn: 0.2681348	total: 4.11s	remaining: 4m 49s
7:	learn: 0.2569842	total: 4.58s	remaining: 4m 41s
8:	learn: 0.2478066	total: 5.07s	remaining: 4m 36s
9:	learn: 0.2410701	total: 5.57s	remaining: 4m 33s
10:	learn: 0.2357826	total: 6.05s	remaining: 4m 28s
11:	learn: 0.2302520	total: 6.55s	remaining: 4m 26s
12:	learn: 0.2265553	total: 7.04s	remaining: 4m 23s
13:	learn: 0.2237080	total: 7.49s	remaining: 4m 20s
14:	learn: 0.2209691	total: 7.95s	remaining: 4m 17s
15:	learn: 0.2176854	total: 8.41s	remaining: 4m 14s
16:	learn: 0.2153899	total: 8.87s	remaining: 4m 11s
17:	learn: 0.2135348	total: 9.33s	remaining: 4m 9s
18:	learn: 0.2116383	total: 9.78s	remain

157:	learn: 0.1409773	total: 1m 16s	remaining: 2m 46s
158:	learn: 0.1408610	total: 1m 17s	remaining: 2m 45s
159:	learn: 0.1406999	total: 1m 17s	remaining: 2m 45s
160:	learn: 0.1405787	total: 1m 18s	remaining: 2m 44s
161:	learn: 0.1403404	total: 1m 18s	remaining: 2m 43s
162:	learn: 0.1401237	total: 1m 19s	remaining: 2m 43s
163:	learn: 0.1399576	total: 1m 19s	remaining: 2m 42s
164:	learn: 0.1398417	total: 1m 20s	remaining: 2m 42s
165:	learn: 0.1396628	total: 1m 20s	remaining: 2m 41s
166:	learn: 0.1394324	total: 1m 20s	remaining: 2m 41s
167:	learn: 0.1392712	total: 1m 21s	remaining: 2m 40s
168:	learn: 0.1390109	total: 1m 21s	remaining: 2m 40s
169:	learn: 0.1388423	total: 1m 22s	remaining: 2m 39s
170:	learn: 0.1384974	total: 1m 22s	remaining: 2m 39s
171:	learn: 0.1383837	total: 1m 23s	remaining: 2m 38s
172:	learn: 0.1382817	total: 1m 23s	remaining: 2m 38s
173:	learn: 0.1381564	total: 1m 24s	remaining: 2m 37s
174:	learn: 0.1379561	total: 1m 24s	remaining: 2m 37s
175:	learn: 0.1377852	total:

310:	learn: 0.1212715	total: 2m 30s	remaining: 1m 31s
311:	learn: 0.1211478	total: 2m 30s	remaining: 1m 30s
312:	learn: 0.1210885	total: 2m 31s	remaining: 1m 30s
313:	learn: 0.1210435	total: 2m 31s	remaining: 1m 29s
314:	learn: 0.1209982	total: 2m 32s	remaining: 1m 29s
315:	learn: 0.1209485	total: 2m 32s	remaining: 1m 28s
316:	learn: 0.1208720	total: 2m 33s	remaining: 1m 28s
317:	learn: 0.1207864	total: 2m 33s	remaining: 1m 27s
318:	learn: 0.1206982	total: 2m 34s	remaining: 1m 27s
319:	learn: 0.1205861	total: 2m 34s	remaining: 1m 26s
320:	learn: 0.1205422	total: 2m 34s	remaining: 1m 26s
321:	learn: 0.1204988	total: 2m 35s	remaining: 1m 25s
322:	learn: 0.1204556	total: 2m 35s	remaining: 1m 25s
323:	learn: 0.1204123	total: 2m 36s	remaining: 1m 24s
324:	learn: 0.1203160	total: 2m 36s	remaining: 1m 24s
325:	learn: 0.1202127	total: 2m 37s	remaining: 1m 23s
326:	learn: 0.1201554	total: 2m 37s	remaining: 1m 23s
327:	learn: 0.1201111	total: 2m 38s	remaining: 1m 22s
328:	learn: 0.1199751	total:

465:	learn: 0.1103269	total: 3m 44s	remaining: 16.4s
466:	learn: 0.1102896	total: 3m 45s	remaining: 15.9s
467:	learn: 0.1101379	total: 3m 45s	remaining: 15.4s
468:	learn: 0.1100638	total: 3m 46s	remaining: 15s
469:	learn: 0.1100344	total: 3m 46s	remaining: 14.5s
470:	learn: 0.1099230	total: 3m 47s	remaining: 14s
471:	learn: 0.1098259	total: 3m 47s	remaining: 13.5s
472:	learn: 0.1097822	total: 3m 48s	remaining: 13s
473:	learn: 0.1096510	total: 3m 48s	remaining: 12.5s
474:	learn: 0.1096233	total: 3m 49s	remaining: 12.1s
475:	learn: 0.1095307	total: 3m 49s	remaining: 11.6s
476:	learn: 0.1094643	total: 3m 50s	remaining: 11.1s
477:	learn: 0.1093613	total: 3m 50s	remaining: 10.6s
478:	learn: 0.1093300	total: 3m 51s	remaining: 10.1s
479:	learn: 0.1092788	total: 3m 51s	remaining: 9.66s
480:	learn: 0.1092082	total: 3m 52s	remaining: 9.18s
481:	learn: 0.1091784	total: 3m 52s	remaining: 8.7s
482:	learn: 0.1090883	total: 3m 53s	remaining: 8.21s
483:	learn: 0.1090196	total: 3m 53s	remaining: 7.73s


Модели LightGBM и CatBoost выдали F1-меру > 0.75, но на наших данных LightGBM выиграла=) Логистическая регрессия была близка к ним!

## Выводы
- В ходе работы мы изучили данные
- Подготовили малую часть данных (1000 строк) с помощью предобученной модели DistilBERT
- Обучили различные модели: логистическую регрессию, LightGBM, CatBoost, и сравнили их F1-меру
- Подготовили данные с помощью TF-IDF и обучили те же модели на новых данных
- Получили F1-меру = 0.77 в модели LightGBM.